---
title: "Starter Notebook"
author: "Skylar Pietz"
format: html
execute:
  keep-md: true
---

```{r setup, include = FALSE}
#Load in any packages you need
library(tidyverse)

#Read in any data

```

## Getting Started

You've used Quarto Notebooks a few times already, so you're starting to get the hang of this. Start by editing the Notebook `title:` and make sure you credit yourself as the `author:`. As a reminder, R-code appears in grey chunks throughout your notebook. Anything within these grey blocks must be executable R code. When you are outside of an R chunk, in a section of the notebook with a white background, you can type freely as if you were in a typical Word document.

You'll quickly see the value of including text to explain your code chunks, and motivate your analysis. Trying to get help from a notebook that doesn't effectively use these discussions will be quite difficult. Additionally, future *you* will be very thankful to present-day *you* if you document your analyses well, and will be quick to curse present-day *you* when you do not.

Now, remove this *Getting Started* text and you'll be ready to start sciencing.

# Challenge 1

```{r}
# Challenge 1

nucleotides <- c("A", "C", "G", "T")
nucleotides

```

Challenge 1 involved creating a list of the four nucleotides A, C, G, and T, using the c() function.

# Challenge 2

```{r}
# Challenge 2

Genome <- 15 
randGenome <- sample(nucleotides, size = Genome, replace = TRUE)
randGenome



```

```{r}
randGenome <- paste(randGenome, collapse = "")
randGenome

```

In Challenge 2 I used code to create a random string of 15 nucleotides, titled randGenome. The `paste()` method was used to collapse the genome of nucleotides down to a single string, so it's easier to read. When I compared my collapsed genome with a classmate, I noticed that we generated very different genomes.

# Challenge 3

```{r}
# Challenge 3

Genome <- 1500 
randGenome <- sample(nucleotides, size = Genome, replace = TRUE)
randGenome <- paste(randGenome, collapse = "" )
randGenome

```

```{r}
# Challenge 3 continued

set.seed(215)
genomeLength <- 1500

randGenome <- sample(nucleotides, size = genomeLength, replace = TRUE)

randGenome <- paste(randGenome, collapse = "")
randGenome
```

Challenge 3 consisted of using code to generate a random genome of 1500 nucleotides long. Once again, I used the `paste()` method in code to collapse it from a list down to a single string. I also added another line, using set.seed(215) to initialize a random number generator with the seed 215. And once again, the `paste()` function was used to collapse the list down to one single string.

# Challenge 4

```{r}
# Challenge 4

set.seed(215)
Genome <- 100
randGenome <- sample(nucleotides, size = Genome, replace = TRUE)
randGenome <- paste(randGenome, collapse = "")
randGenome

```

Here, I used `set.seed(215)` to generate a random genome (`randGenome`) consisting of 100 nucleotides, collapsed down to a single string with the `paste()` function. I counted the frequency of Adenine (A) in the resulting "genome" above and got 23.

## **`for`** Loops

```{r}
mySum <- 0

for(i in 1:10){
  mySum <- mySum + i
  print(mySum)
}
```

Before starting Challenge 5, I began learning different ways to run a new programming technique called the `for` loop. In programming, a for loop allows us to run a set of instructions over and over for some predetermined number of iterations. Doing so will help me create loops to analyze genomes.

# Challenge 5

```{r}
# Challenge 5 

myProduct <- 1
for(j in 1:15){
  myProduct <- myProduct * j
  print(myProduct)
}
  
```

In Challenge 5, I wrote and executed my own for loop in order to analyze a genome.

# Challenge 6

```{r}
# Challenge 6

set.seed(215)
Genome <- 10
randGenome <- sample(nucleotides, size = Genome, replace = TRUE)
randGenome <- paste(randGenome, collapse = "")
randGenome

```

```{r}

for(j in 1:nchar(randGenome)){
  print(str_sub(randGenome, start = j, end = j))
}

```

## Flow Control

Challenge 6 involved generating a random genome substring consisting of 10 nucleotides and using the `paste()` method to collapse the genome to a single string rather than a list. Next, I wrote a `for` loop to print out each individual nucleotide instead of the entire thing. Next, I added the `1:nchar()` function to run the for loop through all nucleotides in the string. The `str_sub()` function was used to extract individual nucleotides. By using `str_sub()` and using `j` as both the start and end for subsetting, I was able to have the for loop extract each individual nucleotide.

# Challenge 7

```{r}
nucleotides <- c("A", "C", "G", "T")
genomeLength <- 10

randGenome <- paste(
  sample(nucleotides, size = genomeLength, replace = TRUE),
                   collapse = "")
print(randGenome)
```

```{r}
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "A"){
    print(str_sub(randGenome, start = i, end = i))
  }
}
```

```{r}
countA <- 0 
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "A"){
    countA <- countA + 1
    print(countA)
  }
}
```

In Challenge 7, I created a genome 10 nucleotides long using the (`randGenome`) function. I then adapted the for loop in order to count the number of Adenine (A) in randGenome. Using these functions, I was able to determine that there were 3 Adenine.

# Challenge 8

```{r}
countG <- 0 
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "G"){
    countG <- countG + 1
    print(countG)
  }
}
```

```{r}
countC <- 0 
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "C"){
    countC <- countC + 1
    print(countC)
  }
}
```

```{r}
countT <- 0 
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "T"){
    countT <- countT + 1
    print(countT)
  }
}
```

## Vibrio Cholerae Chromosome DNA

```{r}
vib_c <- scan("VibrioCholerae.txt", what = "character", sep = NULL)
```

Challenge 8 was similar to Challenge 7 in that I created a randome nucleotide and used the for loop to count the frequencies of each of the four individual nucleotides. I then was able to perform these functions on a real genome from the *Vibrio Cholerae* chromosome DNA, consisting of 1,108,250 nucleotides. I read this large genome into R with the scan() function.

# Challenge 9

```{r}
countA <- 0 
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "A"){
    countA <- countA + 1
  }
}

print(countA)
```

```{r}
countG <- 0 
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "G"){
    countG <- countG + 1
  }
}

print(countG)
```

```{r}
countC <- 0 
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "C"){
    countC <- countC + 1
  }
}

print(countC)
```

```{r}
countT <- 0 
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "T"){
    countT <- countT + 1
  }
}

print(countT)
```

In Challenge 9, I was able to use the code from Challenge 8 to count the frequency of each nucleotide in the *Vibrio Cholerae* chromosome. After doing this, the frequency of adenine was 293942, Guanine was 256024, Cytosine was 263573, and Thymine was 294711.

# Challenge 10

```{r}
rosalinddata <- scan("rosalind_dna.txt", what = "character", sep = NULL)
```

```{r}
countA <- 0 
for(i in 1:nchar(rosalinddata)){
  if(str_sub(rosalinddata, start = i, end = i) == "A"){
    countA <- countA + 1
  }
}

print(countA)
```

```{r}
countG <- 0 
for(i in 1:nchar(rosalinddata)){
  if(str_sub(rosalinddata, start = i, end = i) == "G"){
    countG <- countG + 1
  }
}

print(countG)
```

```{r}
countC <- 0 
for(i in 1:nchar(rosalinddata)){
  if(str_sub(rosalinddata, start = i, end = i) == "C"){
    countC <- countC + 1
  }
}

print(countC)
```

```{r}
countT <- 0 
for(i in 1:nchar(rosalinddata)){
  if(str_sub(rosalinddata, start = i, end = i) == "T"){
    countT <- countT + 1
  }
}

print(countT)
```

In Challenge 10, I used code once again to scan in another genome from the rosalind website (similar to what I did in the previous challenge with the Vibrio Cholerae genome). I then used the code from Challenge 9 in order to calculate the frequency of each nucleotide in the genome string. I determined the frequency of Adenine in the Rosalind DNA genome to be 223, Guanine was 202, Cytosine was 233, and Thymine was 201. I got the Rosalind Challenge correct!

# Summary

In all, I learned a lot throughout this notebook. Including:

-   Created a list of `nucleotides`

-   Used the `sample()` function to create a random \"genome\" using the characters found in the `nucleotides` list.

-   Used `paste()` with the parameter setting `collapse = ""` to collapse my random genome into a single long string of nucleotides.

-   Used `set.seed()` to set a seed for a random number generator to make my results reproducible over time and across different machines (and different researchers too).

-   Used a for loop to repeat a set of simple instructions for a pre-determined number of iterations.

-   Used programmatic flow control in the form of `if/else if/else` statements to run code only when a particular condition is satisfied.

-   Constructed a `for` loop and used `if` statements to count the frequency of each of the four nucleotides in a genome string.

-   Solved a bioinformatics problem on the *rosalind* platform!
