---
title: "Group 3 Project"
author: "Brianna Nason"
format: html
execute:
  keep-md: true
---

```{r setup, include = FALSE}
#Load in any packages you need
library(tidyverse)

#Read in any data

```


## Getting Started

You've used Quarto Notebooks a few times already, so you're starting to get the hang of this. Start by editing the Notebook `title:` and make sure you credit yourself as the `author:`. As a reminder, R-code appears in grey chunks throughout your notebook. Anything within these grey blocks must be executable R code. When you are outside of an R chunk, in a section of the notebook with a white background, you can type freely as if you were in a typical Word document.

You'll quickly see the value of including text to explain your code chunks, and motivate your analysis. Trying to get help from a notebook that doesn't effectively use these discussions will be quite difficult. Additionally, future *you* will be very thankful to present-day *you* if you document your analyses well, and will be quick to curse present-day *you* when you do not.

Now, remove this *Getting Started* text and you'll be ready to start sciencing.

## Challenge 1


```{r}
Nucleotides <- c("A", "C", "G", "T")
Nucleotides
```


## Challenge 2


```{r}
randGenome <- 15
schedule <- sample(Nucleotides, size = randGenome, replace = TRUE)
schedule
paste(randGenome, collapse = "")
```


## Challenge 3


```{r}
Genome_size <- 1500
randGenome <- sample(Nucleotides, size = Genome_size, replace = TRUE)
randGenome<- paste(randGenome, collapse = "")
randGenome
```

```{r}
set.seed(215)
Genome_size <- 1500
randGenome <- sample(Nucleotides, size = Genome_size, replace = TRUE)
randGenome <- paste(randGenome, collapse = "")
randGenome
```


## Challenge 4


```{r}
set.seed(215)
Genome_size <- 100
randGenome <- sample(Nucleotides, size = Genome_size, replace = TRUE)
randGenome <-paste(randGenome, collapse = "")
randGenome
```


## Loops


```{r}
mySum <- 0
for(i in 1:10){
  mySum <- mySum + i
  print(mySum)
}
```


## Challenge 5


```{r}
myProduct <- 1
for(j in 1:15){
  myProduct <- myProduct + j
  print(myProduct)
}
```


## Challenge 6


```{r}
Genome_size <- 10
randGenome <- sample(Nucleotides, size = Genome_size, replace = TRUE)
randGenome<- paste(randGenome, collapse = "")
randGenome
for(i in 1:nchar(randGenome)){
    print(str_sub(randGenome, start = i, end = i))
  }

```


## Flow Control


```{r}
Nucleotides <- c("A", "C", "G", "T")
Genome_size <- 10

randGenome <- paste(
  sample(Nucleotides, size = Genome_size, replace = TRUE),
                   collapse = "")
print(randGenome)
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "A"){
    print(str_sub(randGenome, start = i, end = i))
  }
}
```


## Challenge 7


```{r}
Nucleotides <- c("A", "C", "G", "T")
Genome_size <- 10

randGenome <- paste(
  sample(Nucleotides, size = Genome_size, replace = TRUE),
                   collapse = "")
print(randGenome)

Adenine_count <- 0
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "A")
    Adenine_count <- Adenine_count + 1
    
    print(Adenine_count)
  }


```


## Challenge 8


```{r}
Nucleotides <- c("A", "C", "G", "T")
Genome_size <- 10

randGenome <- paste(
  sample(Nucleotides, size = Genome_size, replace = TRUE),
                   collapse = "")
print(randGenome)

Adenine_count <- 0
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "A")
    Adenine_count <- Adenine_count + 1
    
    print(Adenine_count)
}

Cytosine_count <- 0
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "C")
    Cytosine_count <- Cytosine_count + 1
    
    print(Cytosine_count)
}

Guanine_count <- 0
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "G")
    Guanine_count <- Guanine_count + 1
    
    print(Guanine_count)
}

Thymine_count <- 0
for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "T")
    Thymine_count <- Thymine_count + 1
    
    print(Thymine_count)
  }
```


## Challenge 9


```{r}
vib_c <- scan("/Users/briannanason/Documents/Bioinformatics info/VibrioCholerae.txt", what = "character", sep = NULL)
```

```{r}
Adenine_count <- 0
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "A")
    Adenine_count <- Adenine_count + 1
}

Cytosine_count <- 0
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "C")
    Cytosine_count <- Cytosine_count + 1
}

Guanine_count <- 0
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "G")
    Guanine_count <- Guanine_count + 1
}

Thymine_count <- 0
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "T")
    Thymine_count <- Thymine_count + 1
}

print(c(Adenine_count, Cytosine_count, Guanine_count, Thymine_count))
```


## Challenge 10


```{r}
vib_c <- scan("/Users/briannanason/Downloads/rosalind_dna.txt", what = "character", sep = NULL)
```

```{r}
vib_c <- "TGAATCGCCATAGACCAGGTAGCACCACGGATCGACCGTGCAAGTCCCACATATACCATGTGTTAGTGAAATATTGCGCTTAGGTCCGCAGATAAGAGTGTTTGACAAGGCGGGTAAATACAATATCAGAATCAAGTCGGCTAAACGAGCTCATGATTGGACTGCTTGAACAGTTTAGCGAGTTGCGGGTGACTTACCTGTACGGTGCACGCGTTTTCGTGGAGAACTTTAAGCCCTCGCCAACGTGTTGTAATAGCCTCCTGCTCACCGGCCTCCTATATCCAGACATTGCGACTAGTGCGCATACACCTACTCTTATTTATAAGAACATAAATTGCGCTAACCGTAATGCCTAACACTATGTTCGTGGCTTCGGGTACCAAGGTACCGATGGCACGGGTCGCAACCCGATACGAGCTCTATTGTGGATACTAACTCTGTGATCGTCGCGACAAGATTTGCGAGATTGGGTGGCGTTGAGCAGCAGGCCCCGAACGGTAAACTTACTTGCACGCGATTGGACAAGGAAGTAAATACGCGACTTGAACCCTTATAGAACTTTTAGAGTCGGAATGAGTTATCGAGTCCCAATGGGCTTGTATCTGGGCTGGTTGTCACCTACGCTTCCGAAGACAACATCTAGATTGAGATCAGGTGGTGAGTCTTACACCAATGATTGATACGCTTCACTTGGCGGGCGAGTCCTTGGATAGCTCGCTGCGTTAGCGATTTTCACCCAGTCAGTATTGATTCATTAACTCGGGCGAACTAGATCACTCCCTGAGTAGGGCGCGGACACGCGGTATTGCTGCTAGGGGTGCGTGAACGGAACCACTG"

Adenine_count <- 0
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "A")
    Adenine_count <- Adenine_count + 1
}

Cytosine_count <- 0
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "C")
    Cytosine_count <- Cytosine_count + 1
}

Guanine_count <- 0
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "G")
    Guanine_count <- Guanine_count + 1
}

Thymine_count <- 0
for(i in 1:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "T")
    Thymine_count <- Thymine_count + 1
}

print(c(Adenine_count, Cytosine_count, Guanine_count, Thymine_count))
```


## Origin Replication I

## Challenge 1


```{r}
nucleotide_frequency <- function(genomeString, nucleotide = "A"){
  count <- 0
  for(i in 1:nchar(genomeString)){
    if(str_sub(genomeString, start = i, end = i) == nucleotide){
      count <- count + 1
    }
  }
  return(count)
}

nucleotide_frequency("ACTTGCGGGTATCGAG", "G")
```

```{r}
Nucleotides
Sample_genome <- sample(Nucleotides, size=2000, replace=TRUE)
Sample_genome <- paste(Sample_genome, collapse="")
nucleotide_frequency(Sample_genome, "C")
```


## Challenge 2


```{r}
rand_genome <- function(k){
  Nucleotides <-c ("A", "T", "G", "C")
  rand_genome <- sample(Nucleotides, size=k, replace = TRUE)
  rand_genome <- paste(rand_genome, collapse="")
  return(rand_genome)
}
rand_genome(k=15)
```


## Challenge 3


```{r}
generate_3_mers <- function(myString) {
  list_3_mers <- c()

  for(i in 1:(nchar(myString) - 2)){
  list_3_mers <- list_3_mers %>%
  append(str_sub(myString, start = i, end = i + 2))
    }
  return(list_3_mers)
}

myGenome <- rand_genome(10)
myGenome
generate_3_mers(myGenome)
```


## Challenge 4


```{r}
generate_k_mers <- function(genomeString, k) {
  list_k_mers <- c()

  for(i in 1:(nchar(genomeString) - (k-1))){
  list_k_mers <- list_k_mers %>%
  append(str_sub(genomeString, start = i, end = i + k-1))
    }
  return(list_k_mers)
}

myGenome <- rand_genome(10)
myGenome
generate_k_mers(myGenome, 6)
```


## Challenge 5


```{r}
count_pattern <- function(genomeString, pattern){
  count <- 0
  for(i in 1:nchar(genomeString)){
    if(str_sub(genomeString, start = i, end = i+1) == pattern){
      count = count + 1
    }
  }
  return(count)
}
  
count_pattern(myGenome, "AT")
```


## Challenge 5

Rosalind Challenge


```{r}
rosalind_string <- "TGCTGAACTTTATGCTGAATGCTGAATGCTGAATGCTGAATTGCTGAACCTTGCTGAATGCTGAAATGCTGAAGTATGCTGAATGCTGAAGTGCTGCTGAATGCTGAATGCTGAATGCTGCTGAATGCTGAACATGCTGAACTGCTGAAGTGTCATGCTGAAGGCGATAGGCTGCTGAATTCTTTGCTGAAGTGCTGAATGCTGAAATGCTGAACCTGCTGAAAGTGCTGAATGCTGAATCTTATGCTGAAATGCTGAATGCTGAATGCTGAATGTGCTGAATTTGCTGAATGCTGAACATGCTGAATCCAGGTCTGCTGAAGGTGCTGAAGAGAATCACTGCTGAAAATTGCTGAAACTGTTAGCTGCTGAAATTTGCTGAATGCTGAATGCTGAATCGCCTGCTGAATTATGCTGAATGATGCTGAATCGTGCTGAAGTGCTGAATGCTGAAACATACTGCTGAACGTGCTGAATGCTGAACCCATGCTGAATTGCTGAAACCTGCTGAAACTGCTGAAAGTGCTGAATGCTGAATGCTGAAAGACTGCTGAATGCTGAAATGCTGAAGTGCTGAACACGTGCTGAATGCTGAACGTGCTGAATGCTGAAAATCTGCTGAATGCTGAAGTGCTGAATGCTGAATGCTGAATTGCTGAACAATTGCTGAATGCTGAACTAATGTTTCTGCTGAAGCTAACTGCTGAATGCTGAAGTCGAATCGAACTTGCTGAATCAGGTGCTGAATGATCTCATCTGCTGAACCCGTGCTGCTGAATGTGCTGAACGGATGCTGAAAATGTGCTGAACCGCCACTGCTGAAAAGTGCTGAAAACAGTGCTGAAGGCTGCTGAATTGCTGAATGCTGAATGCTGAAGTGCTGAAATGGGATATGCTTGCTGAAATGCTGAAGAGCATGCTGAATGCTGAATGCTGAAATTGCTGAAATATTCTACATACATACGTGCTGAATCTTGCTGAATGCTGAATCGGTGCTGAACTATGCTGAACTGCTGAATGCTGAATCAATTGCTGAA"
count_pattern <- function(myString, pattern){
  count <- 0
  for(i in 1:nchar(myString)){
    if(str_sub(myString, start = i, end = i+8) == pattern){
      count = count + 1
    }
  }
  return(count)
}
  
count_pattern(rosalind_string, "TGCTGAATG")
```

```{r}
count_pattern <- function(myString, pattern){
  count <- 0
  k <- nchar(pattern)
  for(i in 1:nchar(myString)){
    if(str_sub(myString, start = i, end = i+k-1) == pattern){
      count = count + 1
    }
  }
  return(count)
}
```


## Origin Replication II

## Challenge 1


```{r}
find_frequent_kmers <- function(genome, k){
  #Get the Kmers
  kmers <- generate_k_mers(genome,k)
  kmers <- unique(kmers)
  #count occurrences
  kmer_counts <- rep(0,length(kmers))
  for(i in 1:length(kmers)){
    kmer_counts[i] <- count_pattern(genome, kmers [i])
  }
max_freq <- max(kmer_counts)
freq_kmers <- kmers [kmer_counts= max_freq]
return(freq_kmers)
}

find_frequent_kmers("ACACAGACATCCCACCCC",3)
```

```{r}
generate_k_mers <- function(genomeString, k) {
  list_k_mers <- c()

  for(i in 1:(nchar(genomeString) - (k-1))){
  list_k_mers <- list_k_mers %>%
  append(str_sub(genomeString, start = i, end = i + k-1))
    }
  return(list_k_mers)
}
generate_k_mers("ACACAGACATCCCACCCC",3)
```


## Rosalind Challenge

sample problem:


```{r}
find_frequent_kmers <- function(genome, k){
  #Get the Kmers
  kmers <- generate_k_mers(genome,k)
  kmers <- unique(kmers)
  #count occurrences
  kmer_counts <- rep(0,length(kmers))
  for(i in 1:length(kmers)){
    kmer_counts[i] <- count_pattern(genome, kmers [i])
  }
max_freq <- max(kmer_counts)
freq_kmers <- kmers[kmer_counts==max_freq]
return(freq_kmers)
}

find_frequent_kmers("ACGTTGCATGTCGCATGATGCATGAGAGCT",4)
```


Frequent Words Problem:


```{r}
find_frequent_kmers <- function(genome, k){
  #Get the Kmers
  kmers <- generate_k_mers(genome,k)
  kmers <- unique(kmers)
  #count occurrences
  kmer_counts <- rep(0,length(kmers))
  for(i in 1:length(kmers)){
    kmer_counts[i] <- count_pattern(genome, kmers [i])
  }
max_freq <- max(kmer_counts)
freq_kmers <- kmers[kmer_counts==max_freq]
return(freq_kmers)
}

find_frequent_kmers("TTGATCTAGTGCGGCATTGATCTTCGGCTATCGGCTATTGATCTACATGTGAGTGCGGCAAGTGCGGCATACTGTCATCGGCTATACTGTCATTGATCTTTGATCTACATGTGACATGTGTCGGCTATTGATCTTCGGCTATACTGTCAACATGTGTTGATCTTCGGCTATTGATCTAGTGCGGCATTGATCTTCGGCTATTGATCTTCGGCTATTGATCTTCGGCTAACATGTGTCGGCTATACTGTCAAGTGCGGCATACTGTCATACTGTCATTGATCTAGTGCGGCAACATGTGTCGGCTATTGATCTTTGATCTTACTGTCAAGTGCGGCAAGTGCGGCATTGATCTACATGTGTTGATCTAGTGCGGCATTGATCTTACTGTCAAGTGCGGCATCGGCTATCGGCTATACTGTCATACTGTCATTGATCTTTGATCTTACTGTCATCGGCTATTGATCTTACTGTCATCGGCTAACATGTGTCGGCTATACTGTCAACATGTGTTGATCTTTGATCTTTGATCTAGTGCGGCATCGGCTATACTGTCATACTGTCAACATGTGTACTGTCATACTGTCAAGTGCGGCATTGATCTTACTGTCATTGATCTTTGATCTAGTGCGGCATACTGTCATTGATCTAGTGCGGCATTGATCTTCGGCTAACATGTGACATGTGAGTGCGGCATCGGCTATTGATCTAGTGCGGCATTGATCTACATGTGTACTGTCAAGTGCGGCAAGTGCGGCAAGTGCGGCATTGATCTTCGGCTAAGTGCGGCAACATGTG",11)
```


## Challenge 2


```{r}
reverse_complement <- function(substring){
  #reverse the substring
  substring_list <- str_split(substring,"")[[1]]
  rev_substring <- rev(substring_list)
  #Get complimentary Base-Pairs
  rev_complement <- c()
  for(i in 1:length(rev_substring)){
    if(rev_substring[i]=="A"){
      rev_complement <- append(rev_complement, "T")
    }
    if(rev_substring[i]=="T"){
      rev_complement <- append(rev_complement, "A")
    }
    if(rev_substring[i]=="G"){
      rev_complement <- append(rev_complement, "C")
    }
    if(rev_substring[i]=="C"){
      rev_complement <- append(rev_complement, "G")
    }
  }
  rev_complement <- paste(rev_complement, collapse="")
  return(rev_complement)
} 
reverse_complement("AAAACCCGGT")
```

```{r}
rosalind_genome <- scan("/Users/briannanason/Downloads/rosalind_ba1c-2.txt", what = "character", sep = NULL)
```

```{r}
reverse_complement(rosalind_genome)
```


## Origin Replication III

## Challenge 1


```{r}
initialize_k_mer_dict <- function(k){
  nucleotides <- c("A", "C", "G", "T")

  k_mers_dict <- expand.grid(rep(list(nucleotides), k)) %>%
    unite("k_mers", everything(), remove = TRUE, sep = "") %>%
    unique() %>%
    mutate(count = 0)
  
  return(k_mers_dict)
}

k_mers_dict <- initialize_k_mer_dict(9)
```

```{r}
Clump_finding <- function(genome, L, K, t){
  initial_window <- str_sub(genome, i, L)
  my_kmer_dict <- intialize_kmer_dict (K)
  for(i in 1: L-K+1){
    curr_kmer<- str_sub(initial_window, i, i+K-1)
    curr_count<- count_pattern(initial_window, curr_kmer)
    curr_row<- which(my_kmer_dict$kmer==curr_kmer)
    my_kmer_dict$count[curr_row]<-curr_count
  }
  candidates<- my_kmer_dict%>%
    filter(count >=t)%>%
    pull(k_mers)
  
for(j in 2:(ng - (L - 2))){
  rem_pattern <- str_sub(genome, j -1 + (k-1))
  add_pattern <- str_sub(genome, j + (L - 2) - (k - 1), j + (L - 2))
  
  rem_row <- which(kmer_dict$k_mers == rem_pattern)
  add_row <- which(kmer_dict$k_mers == add_pattern)
  
  kmer_dict$count[rem_row] <- kmer_dict$count[rem_row] - 1
  kmer_dict$count[add_row] <- kmer_dict$count[add_row] + 1
  
  if(kmer_dict$count[add_row] >= t){
    candidates <- append(candidates, add_pattern)
    candidates <- unique(candidates)
  }
}
  return(candidates)
}

```


## Challenge 2

