<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.280">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Ainsley Owens">

<title>Starter Notebook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="Replication_Ainsley_Owens_files/libs/clipboard/clipboard.min.js"></script>
<script src="Replication_Ainsley_Owens_files/libs/quarto-html/quarto.js"></script>
<script src="Replication_Ainsley_Owens_files/libs/quarto-html/popper.min.js"></script>
<script src="Replication_Ainsley_Owens_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Replication_Ainsley_Owens_files/libs/quarto-html/anchor.min.js"></script>
<link href="Replication_Ainsley_Owens_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Replication_Ainsley_Owens_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Replication_Ainsley_Owens_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Replication_Ainsley_Owens_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Replication_Ainsley_Owens_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Starter Notebook</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Ainsley Owens </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="intro-to-r-notebook" class="level3">
<h3 class="anchored" data-anchor-id="intro-to-r-notebook">Intro to R Notebook</h3>
<p>In this notebook, I will be exploring several functions of R and learning tools to complete the Replication Origin Notebooks. This will include using the arrow operator <code>&lt;-</code>to store objects in variables, using Râ€™s <code>sample()</code> function to create random genomes for testing functions, and building functions to process and analyze genetic data. I will also be learning about for loops and reading in data from outside sources. These tools will be used to analyze genomes in future work.</p>
</section>
<section id="challenge-1" class="level3">
<h3 class="anchored" data-anchor-id="challenge-1">Challenge 1</h3>
<p>Here I am using the <code>c()</code> function to put four nucleotides (A, G, C, T) into a list.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>Nucleotides <span class="ot">&lt;-</span><span class="fu">c</span> (<span class="st">"A"</span>,<span class="st">"G"</span>,<span class="st">"C"</span>,<span class="st">"T"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>Nucleotides</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A" "G" "C" "T"</code></pre>
</div>
</div>
</section>
<section id="challenge-2" class="level3">
<h3 class="anchored" data-anchor-id="challenge-2">Challenge 2</h3>
<p>Now, I will create a random string of 15 nucleotides, titled randGenome, using the <code>sample()</code> tool. Using the <code>&lt;-</code> will store the list as randGenome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>genome <span class="ot">&lt;-</span> <span class="dv">15</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">sample</span>(Nucleotides, <span class="at">size=</span> genome, <span class="at">replace=</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "A" "T" "A" "A" "G" "A" "T" "A" "G" "C" "A" "T" "C" "T" "C"</code></pre>
</div>
</div>
<p>Now, I will <code>collapse()</code> the genome to create a single string and make it easier to read.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span> (randGenome, <span class="at">collapse=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ATAAGATAGCATCTC"</code></pre>
</div>
</div>
</section>
<section id="challenge-3" class="level3">
<h3 class="anchored" data-anchor-id="challenge-3">Challenge 3</h3>
<p>Now, I will create a random genome that is 1500 nucleotides long, using the <code>sample()</code> function again, naming this genome randGenome. I used the <code>paste()</code> function to collapse the list into a single string.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>genome <span class="ot">&lt;-</span> <span class="dv">1500</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">sample</span>(Nucleotides, <span class="at">size=</span> genome, <span class="at">replace=</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span> (randGenome, <span class="at">collapse=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AGTGCAATACGATAAACAGAAGGATAGAGCATACTCGTGCTCGGCCAGGAGATAGTGAAGGTTGCTTAGGTGGATTCGATAGTCTCCTCCAATAATCCAGTCAGGAGCACCAAATTACAGCCAACTCGCAAGAATGCTGGCATGTAATTGGAGTTAGCGCGACGACCAGAAGCTAGTCTTAGAATCTCCGGCGCCGTGACTAATTTCGCATTCGATTAGCCAAAACGATAAATACCTGAAGGCGACTTACCGTGATAATTGAACGAGAGCGCTTGCACCGGGTCGTCCTGTTTTCGCCTATCGTCAACCGCACTAGCTGTCTAGCACAGACGGGGCATGTGCTAGCCTACCATGCCGTGGTAAAACAAAACCTAAGCAGTGGATTAAGTGGCTACACTCTCTGTGTACGATCTACCGTCTTTGTAGATGCAAGTGAAGTGTATGCCTAATTCCACTCTCTACCGCCGGTCCCTGGATTACTCACGGATTTACCGGCGAGTGAAGCATACGGGATGGGAGTCCTGCGCATGAGGGGTATGCCCGTTCTAACAATTTTAGAGAAAACATTCGAATTAAGCGGAGCCCCCGACTAAGCATTGCGTGTAGGCAATCTCAGCACGGTAGATCTGCGCCCACATCACGCATGCAACGACTATGGTTTTCACCTTACACCGATCTGGTTGACGGAGTCTCAGACTCCTCTCTCAGATACTAACTCCCACCTGGCCGTAACTCTTTGTACTCGTTTTGTAATGCCAGAAAACTGAGGCCCACTATCTGCCCTCTGCAGATGCCCAGATTATACCTTCTTGGCCGGCGAGGTGTGGTAATAGGGCACATCTCTCAACTTTATTATGTAAAATCCTTGCGTAAGTGATTAATGACGCACACGGCATCTCCCTATGGCTGATCTTACTAGCACTATTGTCGTTGTGCGATACATTGGATCGTATCTTACATCAAAATGAACGTACCGATTTGCATGGGCGTCAATTGACGTTGCCCGTACGATGTCGTGAAGCTCAGCTTAGAGCTTTCGAATAACTTTTTGGCTAGCGTAAGACGAGAAAGATACTAACGCGACAGATGGTGCTCAGACCGGCAGACCCTTTTTGACCTCTAAGCTAAATTGCGAACCATGCTAAATAGATTCTGGGAGCACAACGGCGCGCATAGAATGCAATATGCCAGCGTGGCTATACTTGTAGGAAATTGGCTTAAGTGTACCTGCCAGGTCAATTTAGGCCGTCCAACCGATAACTACCAGCGCAATAGGTAAGCGTAAAGGCTATCCCGGTGCCGTGTCGCACGTAGATCTAGTAAAGGGTAGGGTCCCTCTTAGATATGAAGGTTAAGAAGCCCGGTTTGCATAAGTCCAGTTACGCAATTAAGGCTGACAGTAAATAAACCTCTTATGACATAGTGCAGTAGTCACAGTAAGCAGTACGAAACGGAACTTTTCTTAGTCGCGTGACGCAGCCTAGTCTATTCTGCCGGTTC"</code></pre>
</div>
</div>
<p>Here, I am using <code>set.seed(215)</code> to initialize the random number generation with the seed 215, using the <code>paste()</code> function to collapse the list into one string. The <code>set.seed(215)</code> function guarantees that the same random values will be produced each time the code is ran, useful for reproducible results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">215</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>genomeLength <span class="ot">&lt;-</span> <span class="dv">1500</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">sample</span>(Nucleotides, <span class="at">size =</span> genomeLength, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TCCAATGTTTAATCTTGGCAGGTTTCACGGTGAACTTCCAGTAGGCGCCGGGGGCAGGCACACGTAAGTGTAAATCATCGAACGACGCGCTCGGCACTGTGTCTCCAGACCATGAGATTCCATCTCACGGAACATGTGCACATCAAAAAGAACTACCTTGTGTATGCGTCCGTGGAATGATTTTTATGGTTTCGTCGATAGTTATCCGGGTGGAAGGTTCGCAATTCGCCGATTAGCTTGTATCGAGCAAACACCGAGCACATCTATATGAGTGTTGACGTATGTCAGAGACCTACCGATTCCATCTGACGCTAGCAGGCCCCTACCGAAGGGGTTGTCTTCGGGGCGTCCGGCCATCGGAACCTCTTTTAGATGCCTATTTTAAACATCTCAAGTTAACTAAGGTAGTTACGTCAGCCCAACCTCGAGATCTATCTCTCAGTGTAGAGGAACATCGAGTTACCGATGAATAAAACTTCGGCCTTTCTAATGGTTCAGACTAATGCACATAATTAGTTCGCCCGAGATAAGGTAGTGCCTTGCTGGGGGTTACGCTAGCCCCCCCTCCCCAAGGATGACGCTTGCTATCTTACTGGTGGCCGATTTTTCCTGAGCCGGGTGGATCAAATAGATAGTACGCTCGGATTGGCCTGAAATCCGGGTCGCAACCATCGCCTCGCCTAACGCTCTCCCGGATAGTTCCCGACATTCACTTATAACAAAGACCGAAATTCCCAAGTACTCGCCGAGATCTAGACGTGCGAGTTGGTCCCCCCGCAGTAGGCAATTGGTACTATCTGTCACTACGCTAGGCCCGGACGTTTCCCTGGTGTTAGGTTCTAATCCCATGCCGGTTATCCCATCGGATCAGCAGAGTCGCCCACCTTAATAACTCTGGCGCATAAACGGTGCTCCTGCAAATGTCTTAGCTCGTGTTGTCTGATTTCGGATATTCCTAAACTTGAACACGGAGGGTATGTATTCAGTGGTACATGTGGGAAATCTATGGTGGAAGACTAGTTGAGGTCCATCCTAAGGGGAGGTGAATGAGGAGTTACCTGTGGACTAGTTTTCACGGCGGTAGCGCGTGTGCAAATAACGTAACGTACGTAGCCTGGCTCTTACCGCAAAAATCAATGGGTTTGGACGCGTAGGAACTCTACTGCCATCGCCTTCAGTTATCGGTCGGGCCTCACGACAATTTAAATGCGTCTAGGAATGACGAGTAGGAGAACCGCAAGCCCCGACGGAGCAGTCCGCGGAGAACGATCCAACCACTCTGCCCGCAAATGACCCGGCGATTAACGGCTTTTATGAGCCCTAAACACCTGAGATCTCGAGACACGGACATACTGCCGACTCAACCCGAGAAGCGTTTTTGCCGACCAACAGGGTTCGAGGTTTATGAGAGGCGCCCTGATGAGGCCGGAAACGACTGCCGCCTGAAGCTTAAAATTATTCGATCTGTACCTAAACGACCAGGACACACAGCTGCAAATG"</code></pre>
</div>
</div>
</section>
<section id="challenge-4" class="level3">
<h3 class="anchored" data-anchor-id="challenge-4">Challenge 4</h3>
<p>Using <code>set.seed(215)</code>, to ensure the same random sample is created each time I run the code, I created a random genome composed of 100 nucleotides, done by setting the <code>genomeLength</code> to <code>&lt;-100</code>. The <code>paste()</code> function collapsed the list into one single strand.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">215</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>genomeLength <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">sample</span>(Nucleotides, <span class="at">size =</span> genomeLength, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TCCAATGTTTAATCTTGGCAGGTTTCACGGTGAACTTCCAGTAGGCGCCGGGGGCAGGCACACGTAAGTGTAAATCATCGAACGACGCGCTCGGCACTGT"</code></pre>
</div>
</div>
</section>
<section id="challenge-5" class="level3">
<h3 class="anchored" data-anchor-id="challenge-5">Challenge 5</h3>
<p>Here, I am learning how to create a for loop so I can create for loops to analyze genomes. A for loop is a control-flow construct used to run through a dataset and apply the same set of operations on each item in the dataset. To create a for loop, you must set an iterator and provide the instructions to be run each time through the loop. This simple for loop will multiply <code>myProduct</code>. I first initialized the container, <code>myProduct</code>, with a starting value of 1. By setting the <code>for(j in )</code> value to <code>1:15</code>, this instructs the loop to run through the range of integer values from 1 through 15. I then provided the condition on the iterator within a set of parentheses, ending the line with an open bracket. I will indent the next line, as the indented lines following the open bracket are where the instructions to be run through the loop are located.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>myProduct <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>){</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  myProduct <span class="ot">&lt;-</span> myProduct<span class="sc">*</span>j</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(myProduct)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1
[1] 2
[1] 6
[1] 24
[1] 120
[1] 720
[1] 5040
[1] 40320
[1] 362880
[1] 3628800
[1] 39916800
[1] 479001600
[1] 6227020800
[1] 87178291200
[1] 1.307674e+12</code></pre>
</div>
</div>
</section>
<section id="challenge-6" class="level3">
<h3 class="anchored" data-anchor-id="challenge-6">Challenge 6</h3>
<p>Here, I created a random genome consisting of 10 nucleotides. I then created a for loop intended to print each individual nucleotide as opposed to the whole string. I used <code>1:nchar()</code> to run the for loop through all characters in the string. I used the <code>str_sub()</code> function to extract individual nucleotides. The string subset function takes three arguments: the <code>string</code> to be subset, the position to <code>start</code> subsetting, and the position to <code>end</code> subsetting the string. If <code>start</code> and <code>end</code> are the same, a single character will be extracted. By using <code>str_sub()</code> and using <code>j</code> as both the <code>start</code> and <code>end</code> for subsetting, I was able to have the for loop print the individual nucleotides.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">215</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>genomeLength <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">sample</span>(Nucleotides, <span class="at">size =</span> genomeLength, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">paste</span>(randGenome, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>randGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TCCAATGTTT"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(randGenome)){</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span> (<span class="fu">str_sub</span>(randGenome,<span class="at">start=</span> j, <span class="at">end=</span>j))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>         }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "T"
[1] "C"
[1] "C"
[1] "A"
[1] "A"
[1] "T"
[1] "G"
[1] "T"
[1] "T"
[1] "T"</code></pre>
</div>
</div>
</section>
<section id="challenge-7" class="level3">
<h3 class="anchored" data-anchor-id="challenge-7">Challenge 7</h3>
<p>Flow control is a programming concept used to execute code only when certain conditions are satisfied. This can be achieved using <code>if</code>, <code>else if</code>, and <code>else</code> statements. The code in Challenge 6 was adapted to only print occurrences of Adenine (<code>"A"</code>). This is done by using an <code>if</code> statement. I first initialized the container, <code>adenineCount</code>, with a starting value of 0. The <code>for</code> line instructs the code to run through all of the characters in the random genome. The <code>if</code> line instructs the code to only run if the current nucleotide in the random genome is Adenine (<code>"A"</code>). Because I already initialized the container with a starting value of 0, I can instruct the code to increment the value of the container by one every time an Adenine occurs, by following the <code>for</code> line, with <code>adenineCount &lt;- adenineCount +1</code>. After closing the brackets for the flow control, I used the <code>print()</code> function to return the occurrences of Adenine.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>adenineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(randGenome)){</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(randGenome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>){</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    adenineCount <span class="ot">&lt;-</span> adenineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a> <span class="fu">print</span>(adenineCount)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
</section>
<section id="challenge-8" class="level3">
<h3 class="anchored" data-anchor-id="challenge-8">Challenge 8</h3>
<p>Now that I have created a for loop to count the number of occurrences of Adenine in the string, I will adapt the loop to count the frequencies of each of the four nucleotides. I did so by setting containers for each nucleotide, <code>thymineCount</code>, <code>cytosineCount</code>, <code>adenineCount</code>, and <code>guanineCount</code>, to a starting value of zero. I then repeated the code used in Challenge 7, and changed the nucleotide in the <code>if</code> line to count each nucleotide, as well as the <code>nucleotideCount</code> line to the matching nucleotide to increment by 1 each time the nucleotide occurs. After closing the brackets for my loops, I printed each nucleotide count, using the <code>print(c(nucleotideCount))</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>thymineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>cytosineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>adenineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>guanineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(randGenome)){</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(randGenome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"T"</span>){</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>   thymineCount <span class="ot">&lt;-</span> thymineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>   <span class="cf">if</span>(<span class="fu">str_sub</span>(randGenome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>){</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>   adenineCount <span class="ot">&lt;-</span> adenineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>   }</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(randGenome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>){</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>   cytosineCount <span class="ot">&lt;-</span> cytosineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(randGenome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>){</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>   guanineCount <span class="ot">&lt;-</span> guanineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">c</span>(thymineCount, adenineCount, cytosineCount, guanineCount))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5 2 2 1</code></pre>
</div>
</div>
<p>Here, I read in the Vibrio Cholerae genome, composed of 1,108,250 nucleotides, to test my code on a real genome. I named the dataset <code>vib_c</code>, and scanned the data in using the file path in my computer in quotations, followed by <code>, what = "character", sep = NULL</code>. These lines following the <code>scan</code> function must be in parentheses.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>vib_c <span class="ot">&lt;-</span> <span class="fu">scan</span>(<span class="st">"C:/Users/owens/OneDrive/Desktop/VibrioCholerae.txt"</span>, <span class="at">what =</span> <span class="st">"character"</span>, <span class="at">sep =</span> <span class="cn">NULL</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="challenge-9" class="level3">
<h3 class="anchored" data-anchor-id="challenge-9">Challenge 9</h3>
<p>Now I can use the for loop I created to count all of the occurrences of the individual nucleotides in the cholera genome, by replacing <code>randGenome</code> with <code>vib_c</code>, to pull the nucleotides from the cholera genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>adenineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>cytosineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>guanineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>thymineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(vib_c)){</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"T"</span>){</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>   thymineCount <span class="ot">&lt;-</span> thymineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>   <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>){</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>   adenineCount <span class="ot">&lt;-</span> adenineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>   }</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>){</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>   cytosineCount <span class="ot">&lt;-</span> cytosineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>){</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>   guanineCount <span class="ot">&lt;-</span> guanineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">c</span>(thymineCount, adenineCount, cytosineCount, guanineCount))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 294711 293942 263573 256024</code></pre>
</div>
</div>
</section>
<section id="challenge-10" class="level3">
<h3 class="anchored" data-anchor-id="challenge-10">Challenge 10</h3>
<p>Here, I used the <code>scan()</code> function to read in a Rosalind genome, from the Counting DNA Nucleotides problem, to check that my code is working properly. I used the for loop previously created in Challenge 9 to count the occurrences of each nucleotide in the genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>rosalindSequence <span class="ot">&lt;-</span> <span class="fu">scan</span>(<span class="st">"C:/Users/owens/Downloads/rosalind_dna (6).txt"</span>, <span class="at">what =</span> <span class="st">"character"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>I got this Rosalind problem correct, meaning that my code is working properly to count nucleotides.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>rosalindSequence <span class="ot">&lt;-</span> <span class="st">"CTTACAAATGAAGCCATATCCTCCCCCCCACTGGCGCGAAATAGAAACCCAAGCCATGTGTTAGGGCATTATAGGGATTGCGGGGTACGGCTCCTTGGCAACTAATAGATGTGAGGCTCTACACAAGTCCAAGGGCCTTTTCCACTCTAGGCTCATTATAAGCTACCGGCCCCAGTGTATCGTGGGCGCGTACCTGTCAGATAACTTTTCTCCGGGATTTGAGTCGACTCGTGGGTTTAAGATCGGTCTATCCCCCACATCCTGACTGAGGTGCACTTTACAATAATTGGTAGTTTAAACAAGGGCTATGCCCGCCGGCCTCAGACGTTGTTGCAGGAGGCAATGGCCAAAACTGCTTCCAGTAGGTTTTTAGGATCCACTACGACCAAAGCTAGCATTCAGTTGCCTGAGACGCGACTTCCAGTACTTGGCCCCGACGACAGAGGGCTTACAATTGTAGCCCCCCGTTAACCAGGGTTGCACCGGCCCAGATGCCGGCGGATGGTGTAGACTGCGAATGGAGAAACGTATCAGTACCAGGTGATCAAGTGCGAATAGGCAGCGACCTAGATTGGGCAACTTCATCGATCCCATTCGAGGCTCGGGTAGTCGGTTACACCATGGCGTCGCGGTAATGCCCTGCGAACACATTTGCAAGTCTTCGCCAATATACGTGCCTCCCGTGTAATCGGTCGAACCAGAAGCATCAGCCGGATCGGCAGGCGCTAAGCGGTAACCTATCCGCGATTGTTTTCCCAGTTCGTGCTCTTGTTACCCGGATAAATAGAAAGTTATTCATGGGAAGTACGTTCGCAGTCATAACCGCTCGTTAAATTATGTATACTCTTCAGTCAACGCGCGTTTCTACTTGTACATGGTCCCGGGTCCATCTGATGTGGGTCCGTTTAATCTCGAGGTTAGAAGGAGAAACCGATGCCGCCATCAACAAGCG</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="st">"</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>thymineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>adenineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>cytosineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>guanineCount <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">0</span><span class="sc">:</span><span class="fu">nchar</span>(rosalindSequence)){</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalindSequence, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"T"</span>){</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>   thymineCount <span class="ot">&lt;-</span> thymineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>   <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalindSequence, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>){</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>   adenineCount <span class="ot">&lt;-</span> adenineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>   }</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalindSequence, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>){</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>   cytosineCount <span class="ot">&lt;-</span> cytosineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalindSequence, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>){</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>   guanineCount <span class="ot">&lt;-</span> guanineCount <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">c</span>(adenineCount,cytosineCount, guanineCount, thymineCount))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 229 249 242 232</code></pre>
</div>
</div>
<p>Throughout this notebook, I learned several skills. I was able to create a list of nucleotides using <code>c()</code>. Using the <code>sample()</code> function, I was able to use the nucleotide list I provided to create a random genome. The <code>collapse()</code> function allowed me to collapse the genome into one single strand, as opposed to separate single characters. I learned that the <code>set.seed()</code> function allows for the code to set a seed for random number generation to enable reproducible results over time. I learned how to create a for loop to repeat instructions provided, and I used <code>if/else</code> statements to ensure the code is only ran when certain criteria is met. This notebook also taught me how to scan in data from other sources, like the vibrio genome and the Rosalind sequence, using the <code>scan()</code> function. I learned how to correctly solve bioinformatics problems on the Rosalind platform.</p>
</section>
<section id="replication-origin-i" class="level3">
<h3 class="anchored" data-anchor-id="replication-origin-i">Replication Origin I</h3>
<p>In this notebook, I will be working towards developing code that can be used to find the origin of replication in a genome. During DNA replication, proteins must locate the origin where the process begins. Locating specific patterns in the genome are the first step in the process of determining the replication origin. Throughout this notebook, I will be creating and using reusable code.</p>
</section>
<section id="challenge-1-1" class="level3">
<h3 class="anchored" data-anchor-id="challenge-1-1">Challenge 1</h3>
<p>In the first code block, I am reading in the <code>nucleotide_frequency</code> function for further use. There are two arguments, a string (<code>genomeString</code>) and a nucleotide (<code>nucleotide</code>). This function counts the number of occurrences of specific nucleotides in the string. I set the container, <code>count</code>, to a starting value of 0. The <code>for</code> loop instructs the code to run through all characters in the string, and the <code>if</code> statement instructs the code to only run if the nucleotide indicated is present. By setting <code>count</code> to <code>count + 1</code> following the <code>if</code> statement, the code is set to increment by 1 every time the target nucleotide occurs. I am then returning the <code>count</code> of nucleotide occurrences.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>nucleotide_frequency <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeString, <span class="at">nucleotide =</span> <span class="st">"A"</span>){</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genomeString)){</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">str_sub</span>(genomeString, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> nucleotide){</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>      count <span class="ot">&lt;-</span> count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(count)</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Using the <code>sample()</code> function alongside the <code>paste()</code> and <code>collapse""</code>, I generated a random genome with a length of 2000 nucleotides, and stored it as <code>randGenome</code>. Using the <code>nucleotide_frequency()</code> function, I instructed the code to pull from the <code>randGenome</code> sample I just created, and indicated the target nucleotide to count as <code>nucleotide="C"</code> to count the occurrences of Cytosine.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>  genome <span class="ot">&lt;-</span> <span class="dv">2000</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">sample</span>(Nucleotides, <span class="at">size=</span> genome, <span class="at">replace=</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>randGenome <span class="ot">&lt;-</span> <span class="fu">paste</span> (randGenome, <span class="at">collapse=</span><span class="st">""</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nucleotide_frequency</span>(randGenome,<span class="at">nucleotide=</span><span class="st">"C"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 499</code></pre>
</div>
</div>
</section>
<section id="challenge-2-1" class="level3">
<h3 class="anchored" data-anchor-id="challenge-2-1">Challenge 2</h3>
<p>In this challenge, I created a function to compute random genomes. This will make future work in this notebook easier because I no longer need to create genome lists, but instead can pull using the <code>rand_genome()</code> function. I provided the function with the nucleotide list to use, denoted the size parameter as k, and used the <code>paste()</code> and <code>collapse''</code> functions to create one string. To do so, I just need to insert <code>rand_genome()</code> and place the desired genome length in the parentheses, using the <code>k</code> parameter.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>rand_genome <span class="ot">&lt;-</span> <span class="cf">function</span>(k){</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  nucleotides <span class="ot">&lt;-</span><span class="fu">c</span> (<span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  rand_genome <span class="ot">&lt;-</span> <span class="fu">sample</span>(nucleotides, <span class="at">size=</span> k, <span class="at">replace=</span> <span class="cn">TRUE</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  rand_genome <span class="ot">&lt;-</span> <span class="fu">paste</span>(rand_genome, <span class="at">collapse=</span> <span class="st">""</span>)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(rand_genome)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rand_genome</span>(<span class="at">k=</span><span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CTCCCCTATCCTAGC"</code></pre>
</div>
</div>
</section>
<section id="challenge-3-1" class="level3">
<h3 class="anchored" data-anchor-id="challenge-3-1">Challenge 3</h3>
<p>For this challenge, I will create a for loop that can run all the characters in a genome and count consecutive characters. This for loop will count and return all the substrings of 3 nucleotides in the genome, known as 3-mers. I used the <code>rand_genome()</code> function within the new <code>generate_3_mers</code> function that I created to generate a 2000 nucleotide sequence. The <code>generate_3_mers</code> function then reported all of the 3-mers in the genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>generate_3_mers <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeString) {</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  list_3_mers <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(<span class="fu">nchar</span>(genomeString) <span class="sc">-</span> <span class="dv">1</span>)){</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  list_3_mers <span class="ot">&lt;-</span> list_3_mers <span class="sc">%&gt;%</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">append</span>(<span class="fu">str_sub</span>(genomeString, <span class="at">start =</span> i, <span class="at">end =</span> i <span class="sc">+</span> <span class="dv">2</span>))</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list_3_mers)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>list_3_mers <span class="ot">&lt;-</span> <span class="fu">generate_3_mers</span>(<span class="fu">rand_genome</span>(<span class="at">k=</span><span class="dv">2000</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To check if the function collected only 3-mers, I used the <code>tail()</code> function to print the last 25 substrings collected, using <code>n=</code>. I was able to call the <code>list_3_mers</code> data because I named the list in the previous code block. The tail function showed that the last substring was a 2-mer, not a 3-mer (â€œAAâ€).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(list_3_mers, <span class="at">n=</span> <span class="dv">25</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "CCT" "CTT" "TTT" "TTT" "TTA" "TAG" "AGA" "GAT" "ATT" "TTT" "TTG" "TGG"
[13] "GGG" "GGA" "GAG" "AGA" "GAG" "AGT" "GTT" "TTA" "TAA" "AAA" "AAA" "AAA"
[25] "AA" </code></pre>
</div>
</div>
</section>
<section id="challenge-4-1" class="level3">
<h3 class="anchored" data-anchor-id="challenge-4-1">Challenge 4</h3>
<p>For this challenge, I am adapting the function from the previous challenge to create a function that reports k-mers of any size. I did so by adding the k value. I named the list <code>list_k_mers</code> so I am able to call upon this list during future work. The value next to <code>rand_genome</code> at the bottom of the code block determines how many characters the genome will be comprised of, while the value next to <code>myGenome</code> in the last line of code determines how long the k-mers reported by the function are. If I give the <code>rand_genome function</code> a value of 10, named <code>myGenome</code>, then the genome will be 10 nucleotides long. If I give the <code>generate_k_mers function</code> a value of 5, pulling from <code>myGenome</code>, then the function will report the k-mers that are 5 nucleotides long.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>generate_k_mers <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeString, k) {</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  list_k_mers <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(<span class="fu">nchar</span>(genomeString) <span class="sc">-</span> (k<span class="dv">-1</span>))){</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  list_k_mers <span class="ot">&lt;-</span> list_k_mers <span class="sc">%&gt;%</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">append</span>(<span class="fu">str_sub</span>(genomeString, <span class="at">start =</span> i, <span class="at">end =</span> i <span class="sc">+</span> k<span class="dv">-1</span>))</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list_k_mers)</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>myGenome <span class="ot">&lt;-</span> <span class="fu">rand_genome</span>(<span class="dv">10</span>)</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>myGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TTGATCTCTA"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">generate_k_mers</span> (myGenome, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TTGAT" "TGATC" "GATCT" "ATCTC" "TCTCT" "CTCTA"</code></pre>
</div>
</div>
</section>
<section id="challenge-5-1" class="level3">
<h3 class="anchored" data-anchor-id="challenge-5-1">Challenge 5</h3>
<p>Here, I created a function, <code>count_pattern</code>, to count occurrences of a particular pattern within a genome string. The bottom line of code, <code>count_pattern(myGenome, "")</code> tells R which pattern to count throughout the genome. This is function that will be adapted later in the challenge, so this code block is just the outline. In the <code>if</code> statement, the <code>end = i + 1)</code>, the +1 value can be adjusted to account for a larger pattern. The <code>count_pattern()</code> line can also be adjusted for future use. The <code>(myGenome, "TA")</code> can be adapted to a particular string and pattern, which will be seen next.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>count_pattern <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeString, pattern){</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genomeString)){</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">str_sub</span>(genomeString, <span class="at">start =</span> i, <span class="at">end =</span> i<span class="sc">+</span><span class="dv">1</span>) <span class="sc">==</span> pattern){</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>      count <span class="ot">=</span> count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(count)</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a><span class="fu">count_pattern</span>(myGenome, <span class="st">"TA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
</div>
<p>Here, I read in the Rosalind sequence to test that the code is properly working, naming it <code>rosalindString</code> using the <code>&lt;-</code> arrow tool.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>rosalindString <span class="ot">&lt;-</span> <span class="st">"CGCTGCAATGTGCAATGTTGCAATGTGCAATGAATTATGCAATGAATGCAATGTTGCAATGACTTGCAATGTATGCAATGGTGCAATGTTGCAATGTGCAATGTGCAATGACCATGCAATGTGCAATGTGCAATGAATTGCAATGTGCAATGTGCAATGTGCAATGTATGGCTGCAATGTGCAATGTGCAATGTGCAATGGTGCAATGCCAGTGCAATGGTTAAGTGCAATGAGTCTTGACTGCAATGTTTTGCAATGGTTTTGCAATGTGCAATGCTGCAATGGAATGCAATGTGCAATGCTGCAATGTGCAATGTGCGTAAGTGCAATGTGCAATGATGCAATGCCTGTCCATGCAATGTGCAATGTCGGTGCAATGTCATTATGCAATGTGAAGCGAATATGCAATGATGCAATGGTGCAATGTGCAATGTTGCAATGTGCAATGCAATGCAATGGTGCAATGATTTTGCAATGTTGCAATGCCTATACCTAGTCCATGCAATGTGCAATGGTAAATGCAATGGAGTGCAATGCTATGCCTTGCAATGTTGCAATGTGCAATGCTCCTGCAATGGCATGCAATGTGCAATGGTGCAATGGAACTGCAATGCTGCAATGATCCAGATCCGTGTGCAATGTTTGCAATGTGCAATGAGGGCGTTGCAATGCCGAGAGCTGCAATGGTGCAATGGTGCAATGACAGATGCAATGGTGCAATGGATGCAATGCCTGCAATGCTGCAATGTGCAATGTGCAATGATGCAATGTGCAATGTGCAATGGCCATTGCAATGTGCAATGTGCAATGTGATCAATTGCAATGTGCAATGTTGAATGCAATGCTATTCGGTCTGCAATGTTTGCAATGTGTGCAATGTGTGCAATGCTGCAATGTTTGCTGCAATGAAATATGCAATGCCCTTTGCAATGTGCAATGTTGCAATGATGCAATGGCCCGCTGCAATGAACATGTTGCAATGTGCAATG"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this code block, the outline function is adapted to run the Rosalind sequence. I adjusted the function by changing the if statement to <code>i+8</code>, because the pattern that is being looked for in the sequence is 9 characters long. In the last line of code, <code>count_pattern()</code>, I replaced the string to pull from <code>rosalindString</code>, and I copied in the pattern that is being counted, provided by Rosalind. Copying my solution into Rosalind, I determined that my answer is correct, meaning the code is working properly. (I adjusted this code to be functional in Replication Origin II, changing the <code>end= i+8</code> to <code>end= i+ k -1</code>. Having the <code>end= i+8</code> caused the <code>count_pattern</code> function to not be usable in Challenge 1 of Replication Origin II, because the string was not 9 nucleotides long.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>count_pattern <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeString, pattern){</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  k<span class="ot">&lt;-</span> <span class="fu">nchar</span>(pattern)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genomeString)){</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">str_sub</span>(genomeString, <span class="at">start =</span> i, <span class="at">end =</span> i<span class="sc">+</span>k<span class="dv">-1</span>) <span class="sc">==</span> pattern){</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>      count <span class="ot">=</span> count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(count)</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a><span class="fu">count_pattern</span>(<span class="st">"ATATAT"</span>, <span class="st">"AT"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
</div>
<p>While completing this notebook, I learned several skills and concepts. I learned to use functions to write reusable code, which can be helpful when trying to find similar things in several different genomes. Using functions to write reusable code prevents having to rewrite code several times to complete the same task within different genomes. I have worked more to learn how to understand and break down genomes using code. In this notebook, I learned how to count the number of occurrences of patterns in the genome.</p>
</section>
<section id="replication-origin-ii" class="level3">
<h3 class="anchored" data-anchor-id="replication-origin-ii">Replication Origin II</h3>
<p>In this notebook, I will continue developing code with the goal of finding the replication origin in a genome. DNA replication begins in a certain position on the genome, known as the origin. This origin consists of a set of DNA sequences that signals the cells to start replication.</p>
</section>
<section id="challenge-1-2" class="level3">
<h3 class="anchored" data-anchor-id="challenge-1-2">Challenge 1</h3>
<p>In bacteria, one of the proteins involved in replication is known as DnaA. DnaA binds to DnaA boxes, sequences in the genome, strings of 9 nucleotides that are more frequently repeated around the origin of replication.</p>
<p>In this challenge, I will write code to search for frequently occurring 9-mers. I will begin by using a previously created function to practice counting k-mers. Using the <code>generate_k_mers</code> function I previously created, I read in the sample sequence from the notebook guide and named it <code>myGenome</code>. By changing the value on the last line of code next to <code>myGenome</code>, I instructed R to report the 3-mers in the sequence.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>generate_k_mers <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeString, k) {</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  list_k_mers <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(<span class="fu">nchar</span>(genomeString) <span class="sc">-</span> (k<span class="dv">-1</span>))){</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>  list_k_mers <span class="ot">&lt;-</span> list_k_mers <span class="sc">%&gt;%</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">append</span>(<span class="fu">str_sub</span>(genomeString, <span class="at">start =</span> i, <span class="at">end =</span> i <span class="sc">+</span> k<span class="dv">-1</span>))</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list_k_mers)</span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>myGenome <span class="ot">&lt;-</span> <span class="st">"ACACAGACATCCCACCCC"</span></span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>myGenome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ACACAGACATCCCACCCC"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">generate_k_mers</span> (myGenome, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "ACA" "CAC" "ACA" "CAG" "AGA" "GAC" "ACA" "CAT" "ATC" "TCC" "CCC" "CCA"
[13] "CAC" "ACC" "CCC" "CCC"</code></pre>
</div>
</div>
<p>I will now adapt the <code>generate_k_mers</code> function to create a function that will count and report frequently occurring k-mers. This code block is intended to find and count k-mers in the genome, loop through the list of k-mers and return the most frequently occurring k-mers. By giving <code>k</code> a value of 4, this code will return the most frequent 4-mer. I can change this <code>k</code> value to determine what length of most frequent k-mer is returned.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>frequent_kmers <span class="ot">&lt;-</span> <span class="cf">function</span>(genome, k) {</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  kmers <span class="ot">&lt;-</span> <span class="fu">generate_k_mers</span>(genome, k)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  kmers <span class="ot">&lt;-</span> <span class="fu">unique</span>(kmers)</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  kmer_counts <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, <span class="fu">length</span>(kmers))</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span> <span class="fu">length</span>(kmers)) {</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>    kmer_counts[i] <span class="ot">&lt;-</span> <span class="fu">count_pattern</span>(genome, kmers[i])</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>  max_count <span class="ot">&lt;-</span><span class="fu">max</span> (kmer_counts)</span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>  frequent_kmers <span class="ot">&lt;-</span> kmers[kmer_counts<span class="sc">==</span>max_count]</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(frequent_kmers)</span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a>smallGenome <span class="ot">&lt;-</span> <span class="st">"ATCCATTAT"</span></span>
<span id="cb50-15"><a href="#cb50-15" aria-hidden="true" tabindex="-1"></a><span class="fu">frequent_kmers</span>(smallGenome, <span class="at">k=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ATCC" "TCCA" "CCAT" "CATT" "ATTA" "TTAT"</code></pre>
</div>
</div>
<p>Now, to ensure my code is working properly, I will test it on a Rosalind sequence. This code block is using the Rosalind sample sequence, named <code>rosalindGenome</code>, before downloading the full genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>find_frequent_kmers <span class="ot">&lt;-</span> <span class="cf">function</span>(genome, k) {</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  kmers <span class="ot">&lt;-</span> <span class="fu">generate_k_mers</span>(genome, k)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  kmers <span class="ot">&lt;-</span> <span class="fu">unique</span>(kmers)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>  kmer_count <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, <span class="fu">length</span>(kmers))</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span> <span class="fu">length</span>(kmers)) {</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>    kmer_count[i] <span class="ot">&lt;-</span> <span class="fu">count_pattern</span>(genome, kmers[i])</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>  max_count <span class="ot">&lt;-</span><span class="fu">max</span> (kmer_count)</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>  frequent_kmers <span class="ot">&lt;-</span> kmers[kmer_count<span class="sc">==</span>max_count]</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(frequent_kmers)</span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>rosalindGenome <span class="ot">&lt;-</span> <span class="st">"ACGTTGCATGTCGCATGATGCATGAGAGCT"</span></span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a><span class="fu">find_frequent_kmers</span>(rosalindGenome, <span class="at">k =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GCAT" "CATG"</code></pre>
</div>
</div>
<p>Here, I read in the full Rosalind sequence, as data, and then named the data <code>genomeString</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span><span class="st">"ACCGCAGAGTCCCTATGGGCAGCCGCACGCCCTATGGGACCGCAGAGTACCGCAGAGTTGCCACGGATGCCACGGACCCTATGGGTGCCACGGATGCCACGGACAGCCGCACGTGCCACGGACAGCCGCACGCAGCCGCACGACCGCAGAGTACCGCAGAGTACCGCAGAGTTGCCACGGATGCCACGGACCCTATGGGTGCCACGGACAGCCGCACGTGCCACGGACCCTATGGGGGAGTGGTTGCCACGGACCCTATGGGCAGCCGCACGCAGCCGCACGACCGCAGAGTGGAGTGGTTGCCACGGAACCGCAGAGTGGAGTGGTTGCCACGGACAGCCGCACGTGCCACGGACCCTATGGGGGAGTGGTTGCCACGGACCCTATGGGACCGCAGAGTGGAGTGGTACCGCAGAGTCAGCCGCACGTGCCACGGAACCGCAGAGTACCGCAGAGTGGAGTGGTCCCTATGGGCCCTATGGGCAGCCGCACGGGAGTGGTCAGCCGCACGTGCCACGGATGCCACGGACAGCCGCACGACCGCAGAGTGGAGTGGTACCGCAGAGTCAGCCGCACGCAGCCGCACGTGCCACGGAACCGCAGAGTCAGCCGCACGGGAGTGGTGGAGTGGTTGCCACGGAACCGCAGAGTACCGCAGAGTGGAGTGGTCAGCCGCACGACCGCAGAGTTGCCACGGACAGCCGCACGTGCCACGGAACCGCAGAGTCAGCCGCACGTGCCACGGACAGCCGCACGTGCCACGGATGCCACGGACAGCCGCACGTGCCACGGAACCGCAGAGTGGAGTGGTCAGCCGCACGCCCTATGGGGGAGTGGTCAGCCGCACGTGCCACGGACCCTATGGGACCGCAGAGTGGAGTGGTACCGCAGAGT"</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="co">#The genomeString</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>genomeString <span class="ot">&lt;-</span> data[<span class="dv">1</span>]</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>genomeString</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ACCGCAGAGTCCCTATGGGCAGCCGCACGCCCTATGGGACCGCAGAGTACCGCAGAGTTGCCACGGATGCCACGGACCCTATGGGTGCCACGGATGCCACGGACAGCCGCACGTGCCACGGACAGCCGCACGCAGCCGCACGACCGCAGAGTACCGCAGAGTACCGCAGAGTTGCCACGGATGCCACGGACCCTATGGGTGCCACGGACAGCCGCACGTGCCACGGACCCTATGGGGGAGTGGTTGCCACGGACCCTATGGGCAGCCGCACGCAGCCGCACGACCGCAGAGTGGAGTGGTTGCCACGGAACCGCAGAGTGGAGTGGTTGCCACGGACAGCCGCACGTGCCACGGACCCTATGGGGGAGTGGTTGCCACGGACCCTATGGGACCGCAGAGTGGAGTGGTACCGCAGAGTCAGCCGCACGTGCCACGGAACCGCAGAGTACCGCAGAGTGGAGTGGTCCCTATGGGCCCTATGGGCAGCCGCACGGGAGTGGTCAGCCGCACGTGCCACGGATGCCACGGACAGCCGCACGACCGCAGAGTGGAGTGGTACCGCAGAGTCAGCCGCACGCAGCCGCACGTGCCACGGAACCGCAGAGTCAGCCGCACGGGAGTGGTGGAGTGGTTGCCACGGAACCGCAGAGTACCGCAGAGTGGAGTGGTCAGCCGCACGACCGCAGAGTTGCCACGGACAGCCGCACGTGCCACGGAACCGCAGAGTCAGCCGCACGTGCCACGGACAGCCGCACGTGCCACGGATGCCACGGACAGCCGCACGTGCCACGGAACCGCAGAGTGGAGTGGTCAGCCGCACGCCCTATGGGGGAGTGGTCAGCCGCACGTGCCACGGACCCTATGGGACCGCAGAGTGGAGTGGTACCGCAGAGT"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co">#The value of k</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(data[<span class="dv">2</span>])</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>k</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
</div>
<p>Now, I used the <code>find_frequent_kmers</code> function to run through the entire sequence, pulling from <code>genomeString</code>. The code stores the list of most frequently occurring k-mers as <code>frequent_kmers</code>. The desired kmer length to find listed by Rosalind is 14, so I set the <code>k</code> value to equal 14. I used <code>noquote(frequent_kmers)</code> to print the kmers out in a single line with no quotations, and got the Rosalind challenge correct, meaning that my code is working properly.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>frequent_kmers<span class="ot">&lt;-</span> <span class="fu">find_frequent_kmers</span>(genomeString, <span class="at">k=</span><span class="dv">14</span>)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">noquote</span>(frequent_kmers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] CAGCCGCACGTGCC AGCCGCACGTGCCA GCCGCACGTGCCAC CCGCACGTGCCACG CGCACGTGCCACGG
[6] GCACGTGCCACGGA</code></pre>
</div>
</div>
<p>During DNA replication, DNA polymerase can make mistakes, meaning that nucleotides may not appear in DnaA boxes the way we may expect them to. The replication signal that we are looking for to lead us to the origin may appear as the k-mer signaling that replication should begin, or as its reverse compliment. To account for this, we must create a function that counts reverse compliments as well as k-mers.</p>
<p>Adenine and thymine are complementary base pairs, while guanine and cytosine are as well. In double stranded DNA, these complementary base pairs ( A &amp; T, G &amp; C) should be across from each other. If we look across from a segment of DNA, we can see the reverse complement of that segment. This refers to the fact that the complementary sequence is anti-parallel, meaning 5â€™ and 3â€™ ends are oriented in opposite directions on each side of the double strand. Meaning, we read the complementary sequence in reverse order.</p>
</section>
<section id="challenge-2-2" class="level3">
<h3 class="anchored" data-anchor-id="challenge-2-2">Challenge 2</h3>
<p>In this challenge, I will write a function to compute the <code>reverse_compliment()</code> of a <code>genomeSubString.</code></p>
<p>This <code>reverse_compliment</code> function is created to run through the entire sequence that is provided as input, and output the reverse compliment of this sequence. The code is instructed to return the complementary nucleotide each time it loops past a nucleotide, using <code>if</code> statements to instruct the code to run only if the target nucleotide is present. Meaning, every time the code encounters an Adenine, the <code>reverse_compliment</code> function will return Thymine, and vice versa. This is also applied for Cytosine and Guanine, using <code>if</code> statements as well.</p>
<p>I am using the Rosalind Reverse Compliment Problem sample dataset, listing it as <code>substring</code> to ensure that my code is working properly.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>reverse_compliment <span class="ot">&lt;-</span> <span class="cf">function</span>(substring) {</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>  substring_list <span class="ot">&lt;-</span> <span class="fu">str_split</span>(substring, <span class="st">""</span>) [[<span class="dv">1</span>]]</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>rev_substring <span class="ot">&lt;-</span> <span class="fu">rev</span>(substring_list)</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>reverse_compliment <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span> (rev_substring)) {</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(rev_substring[i]<span class="sc">==</span><span class="st">"A"</span>){</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>    reverse_compliment <span class="ot">&lt;-</span> <span class="fu">append</span>(reverse_compliment, <span class="st">"T"</span>)</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(rev_substring[i]<span class="sc">==</span><span class="st">"T"</span>){</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>  reverse_compliment <span class="ot">&lt;-</span> <span class="fu">append</span>(reverse_compliment, <span class="st">"A"</span>)</span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(rev_substring[i]<span class="sc">==</span><span class="st">"C"</span>){</span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a>  reverse_compliment <span class="ot">&lt;-</span> <span class="fu">append</span>(reverse_compliment, <span class="st">"G"</span>)</span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-15"><a href="#cb60-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(rev_substring[i]<span class="sc">==</span><span class="st">"G"</span>){</span>
<span id="cb60-16"><a href="#cb60-16" aria-hidden="true" tabindex="-1"></a>  reverse_compliment <span class="ot">&lt;-</span> <span class="fu">append</span>(reverse_compliment, <span class="st">"C"</span>)</span>
<span id="cb60-17"><a href="#cb60-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-18"><a href="#cb60-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-19"><a href="#cb60-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-20"><a href="#cb60-20" aria-hidden="true" tabindex="-1"></a>reverse_compliment <span class="ot">&lt;-</span> <span class="fu">paste</span>(reverse_compliment, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb60-21"><a href="#cb60-21" aria-hidden="true" tabindex="-1"></a><span class="fu">return</span>(reverse_compliment)</span>
<span id="cb60-22"><a href="#cb60-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-23"><a href="#cb60-23" aria-hidden="true" tabindex="-1"></a>substring <span class="ot">&lt;-</span> <span class="st">"AAAACCCGGT"</span> </span>
<span id="cb60-24"><a href="#cb60-24" aria-hidden="true" tabindex="-1"></a><span class="fu">reverse_compliment</span>(substring)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ACCGGGTTTT"</code></pre>
</div>
</div>
<p>Here, I am again testing to ensure my code is working properly to return the reverse compliment by running the Rosalind Reverse Complement Problem dataset.</p>
<p>Here, I am scanning in the dataset using the <code>scan()</code> function, naming it <code>revcomp_genome</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>revcomp_genome <span class="ot">&lt;-</span> <span class="fu">scan</span> (<span class="st">"C:/Users/owens/Downloads/rosalind_ba1c (1).txt"</span>, <span class="at">what=</span> <span class="st">"character"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that I have scanned in the dataset, and am now running the <code>reverse_compliment</code> function I created through the Rosalind genome. I solved the Rosalind problem correctly, proving that my code is correctly returning the reverse complement of the genome input.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">reverse_compliment</span>(revcomp_genome)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ACTTGTCGTGTCGCTGCCCGGCGGTAACCAGGTGTGTAGCACTGGGCGTATCAACTTAGACTAATAGTGAAGTCTGACAAGAGGTGTAAGGAGAAAAAAACCAGTGAGTATACATGTCTTGCGCAGCGTGGTTTAGAACGAAACCAGCGGGCCTATAGATCTAATTCCGTAATGGTGCAATTCTATACCCTATATCTCGTCAAGGCGGGTACCAGAATTCCCGAGGCTTCCTTGTCCTACTACGGTAGTCATGTGAAGCAGGCCCACTCGCAGTGAGATACCGGCCCGCAGACAGATTAACACCCACGAGCTGGGTGGAGCCGACTCGAGAGTTTTCTATGATCATTAGTATTCATCCTCCCGATGAAGTATGCCATCGTAAACTTTACACCGCACCTTAGTTCAATGTATAATGAGTATTCAGCGGAGGCCATGCTCGTATGTGTTTTGTTTGTATGGATTACCTTCGCGATCTACACTGGCGGCTGTATTGGTACACATCCTCATCCCCGTAGTATAGGAATGAAGTGAGGCTGGTGAGCCTGAAAAAATACCGGAGAGCAAATGTAATAGTGTAGAAAAGCACTGTCCAAAAAATGTCCCTTTTGTGGCGGTTAACATGAAGTTTTAATTCCGGAGATAGTGTTATCTAGGCCGTTTATCCTTCCTTTCGCCTGCTGGTTGACCCCTGCACGCATCTGCTGTACGAACGACACCAGCACAAGAGATCGCCAGGCCTGCGCCTATGTGGGCCCTTGTCTGTGGAAGCGGGTATGGGGAGCCTCACTTTAGTGAGCGTGTCTGAGCGATGCGAATGGCGTTTATTTCGTGTCACACGGGGAATTCTGTGGGGAGTGATTATGGGTAATTTCGAGACTTGGGAGCTGAATGAACCCGAGAGCTCTTTGAGAGAACCTGGTCCAGAGCGCCTACGTGGTCTCACCAGGATTTCTTTCAATGCATGAAAACGCTGAATCAATGCCCTGTCTTCTACTGGTTTTTCAACCTTGATGTGCCGTGTAAAACTAGTTTCAGCAAATGATCCTGGAGTTAAAGATAATTGACACACTTAGTCATTTGGCTGGTGTACATCCTTCAGAAGTTCAGGGGGGTTCGGAATAGGACTGCGAGGCCGAATGATCAGACCCCGAGGATGATCAGGAGAGTACACGTGGCACGGGAACACTGAATTGCCACTCCCGTGGAGACACAGTAATTGGCAACCGTCCAAAATAACATCGTAATGCTTCTGGTTATTATACAATTGCGAGTGTACTGTATTTCCATGGAGAGATCTGCAAAAGCAACGGTTCCGATCTCCCCAATTAACTTAAATTGTATCTGTATGGCCCTCTTCAAAACGATCGCTTCGCAACCCCAGGGTAACAGATGCATCACCCCTTGGGCAGGTCTGAAGATGTCGCAAGTATTTAAAACTGCTCTGCCGAACAAATCGTGGTGTGCTAGTTCGAGGAGTGGGACGGAAGGTACGAGCCAGCGCAAAAAAAGCTCTGGACTCCGCTGCGCCTCTTATACTGATAGGTCCTTCCTGACCTCTCTTTGGATAACACTTTTACTTGCCTCCCCCGTGGACGCATTCCCCCGCAGCTGACACAAAGTAACACACCAAGGAACACGCATGCAACCGATCACTTGTCCGGGACCATTAGCCCGTTGTCAGCACCTGGGGTGGAAAAGGAGACGTAGGACGCATCGGTCCATGATCAAACCACGTTTCAAGAGAGGAAGCCGAGAAAGCTGATGAACGCACCGGAACCACGCGGACTCATTACGACAGCCTAGAACAGTGTTCGTCTCATCTAACGCCCGTCTTTCTACACGTGATTGTAAGATTGGCTGCTGATAGGGCTTTGCCAAACATGTGTAGTCCAACTGCATGATGTCAAAGCTAACGGCGGCTAAAAAACTGGGAATCGTTTCTCTCACTCAGCCGTGCCGTCGTTCGATGACTAAAGGGCTTGCATACATGCGGGCAAGCATTATTAAATGCAATTTACTACGCGTATTTCAGATCACCGAGCGGGCCTAATGTACGCTTATAGTTGAAGTGGTGTTAAGCAAAGGTGATAGCCTTGGTCGTTGAGCGGGACATCTAGTGGACGGGTTAGGAGCACACTAGGAGCACCCAGCTTAAACGCTTAGACGGGCCGGCTTAAGTCCATATTCCGTCTGTACCGAATATGGCAGTGAGAACTCTAGAAAAACACGATTTTGCGTGTACTAAGATAGTGATCCCACAGCAAGGGGCCACTATATAAACTACTTTCTGAATCCCACCATAAGTATCACTAAAGCGAACGTGGCAGTGATCACACCCGTGCTTGATCGTCGTTATGGCTATTCCTACATTGTACTGATAATAGGAATTACCACAGTGTCTTGCGATAGTAACCTCACGAGAGGAAGACACCCGAGCGCTTCATTATCCTAACAAAGAGTGTTATTAGACCAAGTCTCATACCTAGCGGTAATGGTTGCGGCCTAACTTGGCGTGATACAGCGAGGGGCACCCAGTGTGCTCGCTTTTGACTGTTCCCTGTATGTGTGGGACACCGTTCCCAAACGTCCTGAAAGCCCACAGGTACATTTAGTAGCGCATCAGGAAATAATCTCATACCTGTTAAACTCAACATAGAAGCGACCCTAGCTCTGCTTCAGTCTGCTGAGGAATCTCAAAGTGGCATGACCAAAAATGTGCAGTGTCGCTTGATTTCGGCGCGACAGTACGCCAAACACTCCTTCCCGTGAAGACCCGTCTTTGTACAATGCTTCTCGAAAGCGTGAATCTCCTTGGACGTACAAAGCTGTCATTGAGAATGCACACAATAGTTATTGGCCCGGCGAATCAGCATTAAAGGACTTAACAAAAACATCAACAGGCCCGCATTTGAAAACGAAGTGTAGGGACTCTGCAGGACCGTAAAGCGTATCAGAACGGTCGAACTATTGTGTCGGGGGTTTTAAGGGAGGGGAGGACACGTCTTATCGAGCAACGATTTATGCGGTCACGAGAGAGAACACACACCGCAAGTCCACGCTACCATCAATAATTAACTAGGGCTAACCGCACGCATAGTAATGCTAGTTTCGCGTGACTGAGCGTGCAAACTCGACAATGCAGTGCTTAGTGGAGGCCGGGCATAAAAAAGCGACGGACCTCACCGTTGTGGGGTTTTCTCCTCTACCTATTGAATGCTTCTTCTGTACCTTTGAACACACTCCGTCGGTTAGCTCAGAGTTTTGTGGTAAAGAAACATATTTATGATTCCTCAACAACAGTTCGACATATCTATCGTGCTTTTAATTCTTTTTTAAACGGTCCAGGCCTCTTGGCGCTCTGACCGACTAATTTAAATTGTAGCGTGATCACTCACGAGGTCGACATTGGACTGTCGCCCGAGTCCAAGTTATCCAAGCGCGCGCATCACCTCGAAGTTAGCAACGGTACGCTGTGATGTTCAATCTTCGTACGACTCTTTGCCAACAGATTGAGGGAAGGATACCTAAGCACGTCGTATCTGAATGGATAGCTGAAAGCATTTGGGGAACGGATCAGATACTCCATATCAGCGGAACGACTAACCGCCCTATTTATTCCCGATGCCAATATGCGATCATGTTCTACGTCTAAAACTGGTGACTCGTGAATGAGCCAGATAAAGCGCTAATCAACTGTCGGTGCAGGGGTGCAGGCTGAATATGAAACATGAGCTTCGGAAATGGAATCTTCAAAGTTCTTCGCGGAACGGATCTGATACCCTCGCTTCTCTTATGTATTCTATAGGAGGAGCGACCGCGTATCTTAATTGAAGGTTTCAGAGCCATATTGGTCCCTGTCACCACAGTCGCCGGTTGATCGTTCGGGTTAGCACCGATCAATCCTCGGTCCAAAAGGGGCCGAGGTCGATCGCACGAGAGTACCCGCGGTATTCTCTCTCCAATATCGAATCTTGACCTGAACCTGCCGAATCTTTTGTTAGCTTAATCGACCACAACTCGACGCGCACCCTACCCTGCTAGCCGATCCGTCTCCGACGCACTGCTTCAACTTAGACCTGCGGTAGAACTGGTATCAATCTACTTGAGCGACGGTACCCTCGTTACTCCTCCCCAAGGGATGGGCATCTATCGTAGTATAATAGGACCCGATCCACGCCCTTCCCGACAAGTGACCGTCGTCAACTAAAGGAATTGTTTGCGACAATACCCGAGGGGACGGGGGTTAACTCCGACAGATGGGGCACTCGCTTAACGCTTGGTACAGTAGTCCGCCTTACCTGCCAACTGAAGGCGCAAGTTACCGCCTTGTCACAATGACTACAAGGAGTGGGCGCACTGTACCCAAGTTTCCTAGTGTTAGCCGTGTATGCTTGATAGGGGAGCGATCACCTCGTTCGGTTCACCTAAGCACCCGGCAGGATCTGAGCTGTGGTGCTCGTATGTTTGTAGGAACTCCAGTTCGAGCCTCAGATCTGAACAAGCCGAGGGGTGCAACGGAATTATATACTCTCGTTCAAGTTGCCGAGATGGGCCGTGCAGAAGAGGCGGCCGAACTTGTTCGGTTTTCGGTGAACTCCAAAACCGAACGCGACCAACTATTGAAACGTAACACAGTGCTGAATTTGCAAAGCCAATACATTGTACGCCCCCGCGTCGGTCGGAAGCCCCTGTTTAAGATGGATGGGGGTATGAGATGTCCCTGCTAAGAACACAGTAGAATCACGGTGATGGGGCAAGGTGGGCAAAATTTTTCAAATTGTATAGCGCTGTTCGTTTAGACCGGCTAAAATCGCGGCCAGGACAGGGGTATAGAGGACTGTCGTCCTGAAACCTTAAGGGTGTCCCTGCTGTAGCTCACCTAAGAGCGAGCTCTGGCGCGTCCAGTGGTGTTTCTAAATAACTCCAGTCCTCTTAGCGTCATAACACCAAAGGAGGTAGACGAAAGGAACCGTATTTAAACCATCAATGATACCCTCTCCAATCCTTTTTGGCAATCCATGAATTCCAGGTGGTGACCTGCCTCATTGTCTCAGCATCCTGTTTCGCAAACGGTTGTCCATTGGTATAAGAACTGTATCCGCTGTAGCGTGTATTAACAAAAACCGTTGGCATTTGGATAGCCACACTCCCGATGCCTACCTGATGGCGTCGGCCAGGACCTAGCCACGCAGATGCGGTGGAATCCGCAAGTACTGTCTAGCATAGAAGTGGATCTCATGGCCCATTTTAGGTACCTATTCGTGGGCTGTGACTTATCTTAGCCTGAGTAGGCATGATTCGACGCATCCTTTCCTATAGTAATACTTTCTGGAGATGCATATCCTCCTTTTGACAACGGGACTAGGGGACATTTAAAATGTGACCGTGCCGCACCAAGTAACTAACTCATGAAGTGCACCTGTAATTTATCGAATTGTATTCGATGTACCCGCGGAACGCCGGGCGTTTACCACTCCTTCCATCGGCCATTTGAGTGCCCGAAGCAAGTCAAGCTGTGAGTTATCGTCAATTACGTCTTGTACGTTCGTGCAAAAACCAAATGAAGATGAGTTATTACCACGCGCAGTAGGGATTTAGGTGCTGTTCGGGTGTCTTTCTCCGGGAAGAATGACTGGCCTGACATGCTAAACCTTGGGAACCGTAGGCTTTGATGTACGAGCACGCAAGGACGATGTGTGCTATGGTCGTGGGCCTCCACCGGATTGTGGCGACCACCTGCCAGAGAGGTGGACTTAAAAGTTACCACAGAAACGATGACGACTGCGAAAAACTGACATGGAAATTTCGTCGCGCTATGGCAAAGAACCGCGCCAATCACGATCTCACCAAGATTTGCTTCATGGACTGGACCCCCGGAGAGCGGTTGGACTGGGGACAAAAAGGTTTCCAGGACCATGCGCGGACCTAGAGCTACGGTGCGGTTCTCTAAGGGGTTTTTTAACCTCAAGCTTCCTTGACTCGTATCGGATAACCTAACGAGCACAAGTCAGATTTCAAATGTGGCTTCCGGACTTGCAGCTGACACCACCTGGATGGGTCGTCGATAGAAAGCCAGTGACAAGCGTCAGGTCCGCGAACGTTACGGCGGCATGTATTCAGTTTTAAGTAAACCGGCCTGGAGCGTGCCGCGAGTCCGTCTGAGGGATAAGCTTACAGTACAACCTATGAAATATTGCCTATAACCCAGAACATTATTGCTGCCCGATTTAGACCCGCGCTACTTCTTCATCCTTAACCAGTGTTGCTCGGGACTTGTCCTCCAAGTGTCGTCGGCGCAGAAGGTCCAGTCTCTTTACCTCTCACCGTACCGGATACGATAGTTACCTCCCACACCAGGTCCTCTATGCTGGTCAGCGATGTTCGTGATCACGAGAACCCAACTCGTTGTCAGTCCATTTATACACACTCTTGCAGTACTTATGGCATCTACGCTCAGAGAGTAGCTTCATATAACAGACAGGCGTAATCTGGACATGTTTCTCTCGAACGCAGGGGGTTTGCCTTTGAGGGTGTAAACGGCTTCTTGCAGTCGAATTCATAGGACAGGGGCTAGACCCACTAACACGGTTCGAATATACGCCGGGAGGTCTGCACAAAGGTTTATGGCGGCAGGAGTTGCCAGCATCTGTTTCTTGAGGTCAAGCATTGGACCATATATAAAGTCATCCGAGTCGACGAATACGATCGGTTGCACCATCGAATCTCCACCCCAGTGTTCCAAATCTTGCAACGGAGCTAGTTCTAAATCCCCTCGCCCTGGCCTCTGACGAATACGTCGAGGCATTTTTAAGGACCCACGAATCGTGCTCGAGGCGACGGATAGCTCCTCGACCGAGCCGTGACCTTTCGTGATAACGAGGATGGTGGGGCTGGGCATCACCCCAATCTGGGACTGGTCCGACTCAAGTCAAGGAGCTCGATCATCTTACTAGAATGTTGTTACTGATGTGGAGAGTAACACCATTCGAGGTGCATAAGTTCGTTTGGCGGGACAGATTACATCTAATTGTTCGTTGGGCCATACCGTATCAAATTTTTAGCCACGTTGACCGATTTCCGGATAGTTTATCTAAAGGTCATTTGCACCGATATCGAAGAAATGCGCAAGCTGGGAGTGTTTCACAGTACAGGATGGGGGTCCTGCGCTATGTTTTGCTAGGAGAGGCAGAATGTATGGGCAAAGTGCCTATTTCCCCGTGTGAAAGCGCTGTGATCGCATGAGCGTGGAGATCGACTACCAGGAATATACCGGTAGAGTATGGTCTCCACTATTCTGCGGTACGACGCTAAGCTTGTTTAGTCCTGGGATCGAAGCTATTAAAGGTGCGTCCCGTCATGCAGAGTCTCTGACACTCTTGTGCGGGAGTCCTACATTCGGCGAGAGCAGTGTAGAGAGCGCGTTGGCACGGTTGTGATCCCAATGGACTCTCCGTACGTCAGCTATCTCCCTCCCTCAAGGAACCGTCAAGCCGTCGCAAAGAAGCGAAAGGCCCCTGCACGTGTTACTTGGTGAGGAACATGGCCCGCATGGATATTATTTGTCATGTTCCAGAACAGGAACGATGGAGAACCAGAGTCTGGTGAAGATGCCGGGCGCACTGCCTGCCAATCGTAAGACGGTGATCTGCACCCTCCTGCGTAAGGGTGTACTGTGGGTAGGTTAGGTGGTTGGGGCCGAACATTTCCGGAAGAGGTCGAGCTTTGTTGGAGCAGAGCCGTCTCATTAGCAACCGTGCGTACCATGTGCTGCCCAGGCGATGAGTCCAGGCGTCAGTAACAAGCGCGAGTGCGTTGGGTATGCTCTTTGGGCGAGTCCCCCAACCACTCCACACTAGGTGCAAGACTGGTTATAGCCTCGACGGTAGAACGTGCCAGATTCACCTGAAGTAGAAGAGCCATTAAATTCCCGAACGAGTATGCGTGTTAGCGTCTTGTGCGCTGAGAGGTCATGCAGCCAAGCTAGTGTGATCTACTGTGCAAACAGCTAGATAAAGTGCGTGGTTGTTGCGAAATTATCCTGATAGCGTCACGACATTGGGGACGCCCGCTTCCTAAGGTTCGCTTACGGTTTGTCGCCGCAAAGTCGACCACCAATCAGTTATAATAGGTCACCGCTACTAAAAGAGATCATCATCATGAGGATAGCCGCGAAAGTGCTCTGCCGGGAGTTAACGGTGGTTTAGACCTGCACCCCCGAAGAAGAGCCTCGGGATACCCGTACTTAAGTTACTCTTATATTTCACAGTCAGACAGGGATCGTCCTCAGTTTCTCCGTCAGAAATAGTCGGCCGGATAAGCTCTCGAACTAGGCCTCAGCTAGAGCAGTAGGCCGACAAGCTTGCTCACTCATATATGTGGACAATAACCAACCCTAGATCATACCGCAAACTGTCGCGTAAATAGTCATTTTGATTGTACGTCGCACCAAATTCTGTCGACGATGCAAGCGGCATCGCGCTTTAACTAAAACCTCAGTCGGCTGAAAGTCCT"</code></pre>
</div>
</div>
<p>In this notebook, I created code to search a genome for frequent near k-mers and their reverse complements. Finding these helps us get closer to finding the origin of replication, as the code will return the genetic substrings that signal to begin replication to DNA polymerase.</p>
</section>
<section id="replication-origin-iii" class="level3">
<h3 class="anchored" data-anchor-id="replication-origin-iii">Replication Origin III</h3>
<p>In the Replication Origin II notebook, I was able to find k-mers, near k-mers, the most frequent k-mers and their reverse compliments. Scientists have identified that DnaA boxes within different species of bacteria consist of different nucleotide sequences. To find DnaA boxes in certain bacteria, I need to find high-frequency 9-mers within a specific section of the genome.</p>
<p>In this code block, I am creating a dictionary of all possible k-mers, using the <code>initalize_k_mer_dict</code> function and initialized it with a starting value of 0, naming it <code>k_mers_dict</code>, indicating a length of 9.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>initialize_k_mer_dict <span class="ot">&lt;-</span> <span class="cf">function</span>(k){</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  nucleotides <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span>)</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  k_mers_dict <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="fu">rep</span>(<span class="fu">list</span>(nucleotides), k)) <span class="sc">%&gt;%</span></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unite</span>(<span class="st">"k_mers"</span>, <span class="fu">everything</span>(), <span class="at">remove =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>() <span class="sc">%&gt;%</span></span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">count =</span> <span class="dv">0</span>)</span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(k_mers_dict)</span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>k_mers_dict <span class="ot">&lt;-</span> <span class="fu">initialize_k_mer_dict</span>(<span class="dv">9</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this code, I created a <code>clump_finding</code> function. This function pulls from <code>genomeString</code> and takes the parameters <code>L</code>, <code>T</code>, and <code>K</code>. <code>L</code> indicates the length of the window that we are looking at and instructs the code to look in a specific area of the genome. The parameter <code>T</code> counts the frequency of occurrences, while <code>K</code> indicates the length of k-mers we want returned. The first line of code, <code>kmer_dict</code>, initializes the table of k-mers. The <code>for</code> statement tells the code what window to look at, setting up the initial window length. The <code>curr_kmer</code> function instructs the code to float through the initial window and find a pattern, while <code>curr_count</code> counts how many times that pattern appears in the initial window. The <code>curr_row</code> function shows where the k-mer appears in the k-mer dictionary <code>kmer_dict</code>. The <code>candidates</code> function represents possible candidates for DnaA boxes in the initial window. If the count exceeds the t value, the k-mer is added to the list of candidates. Following the <code>candidate</code> function, the code will slide down the sequence one nucleotide at a time until it runs through the rest of the genome. Subtracting <code>L-2</code> from <code>add_pattern</code> prevents the code from running out of the window, reaching the edge of the window without falling off. The <code>rem_pattern</code> and <code>add_pattern</code> functions move through the genome while maintaining the window, adjusting the counts for k-mers. The <code>rem_row</code> function allows us to remove k-mer counts for k-mers that are no longer in the window as we move through the genome. <code>Add_row</code> adds the new nucleotide to the window, moving down the sequence. <code>J</code> represents the left end point of the current window. The final <code>if</code> statement instructs the code to add the pattern to the list of candidates if it exceeds the threshold and updates the list of candidates. The code will run through the entire genome, leaving the last window with a final nucleotide on the right side without falling off the genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>clump_finding <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeString, L, t, k){</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  kmer_dict <span class="ot">&lt;-</span> <span class="fu">initalize_k_mer_dict</span>(k)</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  ng <span class="ot">&lt;-</span> <span class="fu">nchar</span>(genome)</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(L<span class="sc">-</span>(k<span class="dv">-1</span>))){</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>    curr_kmer <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(initial_window, i, i<span class="sc">+</span> (k<span class="dv">-1</span>))</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>    curr_count <span class="ot">&lt;-</span> count<span class="sc">+</span><span class="fu">pattern</span>(initial_window, curr_kmer)</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>    curr_row <span class="ot">&lt;-</span> <span class="fu">which</span>(kmer_dict<span class="sc">$</span>k_mers<span class="sc">==</span>curr_kmer)</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>    kmer_dict<span class="sc">$</span>count[curr_row] <span class="ot">&lt;-</span> curr_count</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>  candidates <span class="ot">&lt;-</span> kmer_dict <span class="sc">%&gt;%</span></span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(count <span class="sc">&gt;=</span>t) <span class="sc">%&gt;%</span></span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(k_mers)</span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>(ng<span class="sc">-</span>(L<span class="dv">-2</span>))){</span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a>    rem_pattern <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, j<span class="dv">-1</span>, j<span class="dv">-1</span><span class="sc">+</span> (k<span class="dv">-1</span>))</span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a>    add_pattern <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, j <span class="sc">+</span> (L<span class="dv">-2</span>)<span class="sc">-</span>(k<span class="dv">-1</span>),j<span class="sc">+</span> (L<span class="dv">-2</span>))</span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a>    rem_row <span class="ot">&lt;-</span> <span class="fu">which</span>(kmer_dict<span class="sc">$</span>k_mers <span class="sc">==</span> rem_pattern)</span>
<span id="cb66-19"><a href="#cb66-19" aria-hidden="true" tabindex="-1"></a>    add_row <span class="ot">&lt;-</span> <span class="fu">which</span>(kmer_dict<span class="sc">$</span>k_mers <span class="sc">==</span> add_pattern)</span>
<span id="cb66-20"><a href="#cb66-20" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb66-21"><a href="#cb66-21" aria-hidden="true" tabindex="-1"></a>    kmer_dict<span class="sc">$</span>count[rem_row] <span class="ot">&lt;-</span> kmer_dict<span class="sc">$</span>count [rem_row] <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb66-22"><a href="#cb66-22" aria-hidden="true" tabindex="-1"></a>    kmer_dict<span class="sc">$</span>count [add_row] <span class="ot">&lt;-</span> kmer_dict<span class="sc">$</span>count [add_row] <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb66-23"><a href="#cb66-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb66-24"><a href="#cb66-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(kmer_dict<span class="sc">$</span>count[add_row] <span class="sc">&gt;=</span> t){</span>
<span id="cb66-25"><a href="#cb66-25" aria-hidden="true" tabindex="-1"></a>      candidates <span class="ot">&lt;-</span> <span class="fu">append</span>(candidates, add_pattern)</span>
<span id="cb66-26"><a href="#cb66-26" aria-hidden="true" tabindex="-1"></a>      candidates <span class="ot">&lt;-</span> <span class="fu">unique</span>(candidates)</span>
<span id="cb66-27"><a href="#cb66-27" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb66-28"><a href="#cb66-28" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb66-29"><a href="#cb66-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(candidates) </span>
<span id="cb66-30"><a href="#cb66-30" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In order to call this function, it must pass a window length, t-value, and k-value. To test our code in class, we scanned in the E. coli genome and attempted to run the code on this genome, discovering that it takes too long. The <code>clump_finding</code> function does work on smaller genomes, but was too slow to run through the 4.6 million nucleotide E. coli genome. This suggests that the code needs to be further optimized to run this genome.</p>
<p>In Replication Origin Notebooks I, II, and III I learned several new coding methods and tools. Including but not limited to creating lists and sample genomes with the <code>c()</code> and <code>sample()</code> functions, as well as using the <code>set.seed()</code> function to set a seed for random number generation to ensure reproducible results. I also learned how to create <code>for</code> loops, and <code>if/else</code> statements to ensure the code is only ran when specific criteria are met. I also learned how to scan in datasets, and correctly solve bioinformatics problems on the Rosalind problem. I learned how to use functions to write reusable code and count the number of occurrences of patterns in a list, or genome in this case. These tools and functions all brought me closer to finding the replication origin of a genome, although future work is still required.</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>