---
title: "Group 3 project"
author: "Saran Ousseini"
format: html
execute:
  keep-md: true
---

# First, we will load our tidyverse

```{r setup, include = FALSE}
#Load in any packages you need
library(tidyverse)

#Read in any data

#
randGenome <- function(k){
  nucleotide <- c("A", "C", "T","G")
  rand_genome <- sample(nucleotide, size = k, replace = TRUE)

  rand_genome <- paste(rand_genome, collapse = "")
  rand_genome 

  return(rand_genome)
}

```

## Challenge 1: In this challenge we will create a list of the four nucleotides A, C, G, and T, using the c() function

```{r}

nucleotide <- c("A", "C", "T", "G")

nucleotide
```

## Challenge 2: We used the code from challenge 1 to create randgenom string of 15 nucleotides. We called the string "randGenoome"

```{r}
numNucleotide <- 15

rand_genome <- sample(nucleotide, size = numNucleotide, replace = TRUE)

rand_genome
```

```{r}
rand_genome <- paste(nucleotide, collapse = "")
rand_genome
```

## Challenge 3: In this challenge, we generated a random genome of 1500 nucleotides long and use the paste() method to collapse it from a list down to a single string. Then printed randGenome to the notebook.

```{r}
set.seed(200)
genomeLength <- 1500

rand_genome <- sample(nucleotide, size = genomeLength, replace = TRUE)

rand_genome <- paste(rand_genome, collapse = "")
rand_genome 
```

## Challenge 4:

In this challenge we set a seed using set.seed(215), then generate a random genome called randGenome, which is made of 100 nucleotide, then we collopased down a single string. We then counted the frequency of Adenine A

```{r}
set.seed(215)
genomeLength <- 100

rand_genome <- sample(nucleotide, size = genomeLength, replace = TRUE)

rand_genome <- paste(rand_genome, collapse = "")
rand_genome 
```

Result: I counted 23 Adenine (A)

```{r}
mySum <- 0

for(i in 1:10){
  mySum <- mySum + i
  print(mySum)
}
```

## Challenge 5:

In this challenge, we initialized a variable called myProduct, and gave it a value of 1. We wrote our own loop, using j (an iterator).

```{r}
myProduct <- 1

for (j in 1:15) {
 myProduct <- myProduct + j
 print(myProduct)
}
```

## Challenge 6:

In this challenge we generated a random genome susbstring consisting of 10 nucleotides and used the past() methode, to collapse our genome to a string instead of a list.

```{r}
nucleotides <- c("A", "C", "G", "T")
genomeLength <- 10

rand_genome <- paste(sample(nucleotides, size = genomeLength, replace = TRUE),
                   collapse = "")
print(rand_genome)
```

## Challenge 7: Adapt the for loop provided above so that it counts the number of occurrences of Adenine (A) in randGenome.

```{r}
for(i in 0:nchar(rand_genome)){
  if(str_sub(rand_genome, start = i, end = i) == "A"){
    print(str_sub(rand_genome, start = i, end = i))
  }
}
```

## Challenge 8:

Here we will adapt a loop we created to count frequencies of all of the nucleotides in the genome string

```{r}
CountA <- 0
CountT <- 0
CountC <- 0
CountG <- 0

my_genome<- randGenome(1000)

for(i in 0:nchar(my_genome)){
  if(str_sub(my_genome, start = i, end = i) == "A"){
    CountA <- CountA + 1}
  
  if(str_sub(my_genome, start = i, end = i) == "T"){
    CountT <- CountT + 1}
  if(str_sub(my_genome, start = i, end = i) == "C"){
   CountC <- CountC + 1}
  if(str_sub(my_genome, start = i, end = i) == "G"){
    CountG <- CountG + 1}  
}
print(c(CountA, CountT, CountC, CountG))
```

## Challenge 9: We used the code from Challenge 8 to count the frequency of each nucleotide in the*Vibrio Cholerae* chromosome.

```{r}
vib_c <- scan("/Users/saranousseini/Desktop/VibrioCholerae.txt", what = "character", sep = NULL)

#vib_c
```

```{r}
CountA <- 0
CountT <- 0
CountC <- 0
CountG <- 0

for(i in 0:nchar(vib_c)){
  if(str_sub(vib_c, start = i, end = i) == "A"){
    CountA <- CountA + 1}
  
  if(str_sub(vib_c, start = i, end = i) == "T"){
    CountT <- CountT + 1}
  if(str_sub(vib_c, start = i, end = i) == "C"){
   CountC <- CountC + 1}
  if(str_sub(vib_c, start = i, end = i) == "G"){
    CountG <- CountG + 1}  
}
print(c(CountA, CountT, CountC, CountG))

```

## Challenge 10: Create A Rosalinda account

## Populate genome given in the Rosalind Challenge

```{r}
genome <- c("CAAACACCAATTGTCTCTTACGAACCGAATTATTTCGTATCAGTTTGGGGCGCGATTACACATGAAGCTTTGTGGAACCTCCGCGTCCAAACGTCAATGTGAAATCCTCAGGGGAGTTAAAACCCCTGACCACGCGGGTCGAGCTCACGTCTTCTGAGGCCCGACTGGTAGGGTCGAAACTTTTAGTAAAACCTACGGTGAATTGACATATTCGTGAGTGAAGTCTACGTAAATTGCTATTCCGTCTGAGCTTTACGATTGTTGAAACTTCAATTTACATTTTAAAATGCATGCAGGTGAGCAGAAGGACATGTATTACGGATGGCACTGATATAACCTTGGTTTCGTGCTCGTTTAAGGCTCATCCATCCTATATTGGTCTATCGGGTCTTCGCTTGATTACAGCTCTTACCTTTTTGAGCCAGCCGGGAAGTAGTTAGATATCGGCCCGACTTAGTTCGAATCTGTGCCCCCGAAGCATGTTACAAATCTGCCCCAGTGTACTACGACAGTATCCGTCAACCTCTGTAGATCCGCTAGATACAATGGGGGATGGAGCCTGGAAATTGGCCCGTAGGGAGGCTTCGTGTGGGTGCACCAATTGCATAGATGTACCAATTCTTGGTATCGCTTAGTTGACCTTGGGCAAAGGACAAGTTACCCATTTATGACGGTGGATGCAGAATAGTAATCAGGCGCAGGCTCATCGAGGACAGGTCTATATCCGTGGGGATAGCTTCATGCGAGGAATTTCAGGTCCGCGCATTATATATTGCGCTTCTACCCGAAGGCGGCTCTTAACTACATTCCGTGCGTGAATAAACCACAGGCAGGCCGCATGTTAC")
```

## Solve problem from Rosalind Challenge

```{r}
CountA <- 0
CountT <- 0
CountC <- 0
CountG <- 0

for(i in 0:nchar(genome)){
  if(str_sub(genome, start = i, end = i) == "A"){
    CountA <- CountA + 1}
  
  if(str_sub(genome, start = i, end = i) == "T"){
    CountT <- CountT + 1}
  if(str_sub(genome, start = i, end = i) == "C"){
   CountC <- CountC + 1}
  if(str_sub(genome, start = i, end = i) == "G"){
    CountG <- CountG + 1}  
}
print(c(CountA, CountC, CountG, CountT))

```

After adding my answer into Rosalind, I was successful in the Rosalind Challenge

# Replication0rigin_PartI

```{r}
nucleotide_frequency <- function(genomeString, nucleotide = "A"){
  count <- 0
  for(i in 1:nchar(genomeString)){
    if(str_sub(genomeString, start = i, end = i) == nucleotide){
      count <- count + 1
    }
  }
  return(count)
}

nucleotide_frequency("ACTTGCGGGTATCGAG", "G")
```

## Challenge 1: In this challenge we will use sample () function in conjunction with the paste (..., collapse ="") to generate a random genome the of at least 2000 .

## Then we used the nucleotide_frequency() function to count the frequency of Cytosine in the random genome you constructed.

```{r}
nucleotide <- c("A", "C", "T", "G")

nucleotide
set.seed(200)
genomeLength <- 2000

randGenome <- sample(nucleotide, size = genomeLength, replace = TRUE)

randGenome <- paste(randGenome, collapse = "")
randGenome 


nucleotide_frequency(randGenome, "C")
```

## Challenge 2: 

```{r}
randGenome <- function(k){
  nucleotide <- c("A", "C", "T","G")
  rand_genome <- sample(nucleotide, size = k, replace = TRUE)

  rand_genome <- paste(rand_genome, collapse = "")
  rand_genome 

  return(rand_genome)
}

randgenome <- randGenome(5)
randgenome
```

```{r}
#myString <- randGenome

str_sub(randGenome(5), start = 1, end = 2)
```

```{r}
generate_2_mers <- function(myString) {
  list_2_mers <- c()

  for(i in 1:(nchar(myString) - 1)){
  list_2_mers <- list_2_mers %>%
  append(str_sub(myString, start = i, end = i + 1))
    }
  return(list_2_mers)
}

generate_2_mers(randGenome(10))
```

## Challenge 3

```{r}
generate_3_mers <- function(myString) {
  list_3_mers <- c()

  for(i in 1:(nchar(myString) - 1)){
  list_3_mers <- list_3_mers %>%
  append(str_sub(myString, start = i, end = i + 2))
    }
  return(list_3_mers)
}

generate_3_mers(randGenome(10))
```

## Challenge 4

```{r}
generate_k_mers <- function(geneomeString, k=3) {
  list_codon <- c()
  for (i in seq(1, nchar(geneomeString) - k +1, by = k)) { 
    list_codon <- list_codon %>%
      append(str_sub(geneomeString, start = i, end = i + k -1))
  }
return(list_codon)
}
generate_k_mers(randGenome(9))
```

```{r}
count_pattern <- function(myString, pattern){
  count <- 0
  for(i in 1:nchar(myString)){
    if(str_sub(myString, start = i, end = i+1) == pattern){
      count = count + 1
    }
  }
  return(count)
}

myString <- c("OUSSEINI")

count_pattern(myString, "OU")
```

## Challenge 5

```{r}
nt_patterns <- function(string, pattern) {
  nt_matches <- 0
  
  for (i in seq(1, nchar(string))){
    if(str_sub(string, i, i + str_length(pattern)-1) == pattern){
      nt_matches = nt_matches + 1
  }
  }
  return(nt_matches)
}
nt_patterns(randGenome(2000), "AGG")
```

# Part II:

## Challenge 1: Find the most frequent words in a string

```{r}
generate_k_mers <- function(geneomeString, k=3) {
  list_codon <- c()
  for (i in seq(1, nchar(geneomeString) - k +1, by = k)) { 
    list_codon <- list_codon %>%
      append(str_sub(geneomeString, start = i, end = i + k -1))
  }
return(list_codon)
}
generate_k_mers("ACACAGACATCCCACCCC")
```

```{r}
find_frequent_kmers <- function(genome, k){
  kmers <- generate_k_mers(genome,k)
  kmers <- unique(kmers)
  kmer_count <- rep(0, length(kmers))
  for (i in 1 : length(kmers)) {
    count_pattern(genome, kmers[i])
    max_count <- max(kmer_count)
    frequent_kmers <- kmers[kmer_count == max_count]
    
  }
  return(frequent_kmers)
}
find_frequent_kmers("ACACAGACATCCCACCCC", k = 3)
```

```{r}
find_frequent_kmers <- function(genome, k){
  kmers <- generate_k_mers(genome,k)
  kmers <- unique(kmers)
  kmer_count <- rep(0, length(kmers))
  for (i in 1 : length(kmers)) {
    count_pattern(genome, kmers[i])
    max_count <- max(kmer_count)
    frequent_kmers <- kmers[kmer_count == max_count]
    
  }
  return(frequent_kmers)
}
find_frequent_kmers("ACGTTGCATGTCGCATGATGCATGAGAGCT", k = 4)
```

## Challenge 2: The Reverse-Complement Problem

```{r}
reverse_complement <- function(substring) {
  substring_list <- str_split(substring, "") [[1]]
  rev_substring <- rev(substring_list)
  reverse_complement <- c()
  for(i in 1:length(rev_substring)) {
    if(rev_substring[i] == "A"){
      reverse_complement <- append(reverse_complement, "T")
    }
    if (rev_substring[i] == "T"){
      reverse_complement <- append(reverse_complement, "A")
    }
    if(rev_substring[i] == "C"){
      reverse_complement <- append(reverse_complement, "G")
    }
    if(rev_substring[i] == "G") {
      reverse_complement <- append(reverse_complement, "C")
    }
  }
  
  reverse_complement <- paste(reverse_complement, collapse = "")
return(reverse_complement)
}
reverse_complement("ACACAGACATCCCACCCC")
```

```{r}
reverse_complement <- function(substring) {
  substring_list <- str_split(substring, "") [[1]]
  rev_substring <- rev(substring_list)
  reverse_complement <- c()
  for(i in 1:length(rev_substring)) {
    if(rev_substring[i] == "A"){
      reverse_complement <- append(reverse_complement, "T")
    }
    if (rev_substring[i] == "T"){
      reverse_complement <- append(reverse_complement, "A")
    }
    if(rev_substring[i] == "C"){
      reverse_complement <- append(reverse_complement, "G")
    }
    if(rev_substring[i] == "G") {
      reverse_complement <- append(reverse_complement, "C")
    }
  }
  
  reverse_complement <- paste(reverse_complement, collapse = "")
return(reverse_complement)
}
reverse_complement("ACGTTGCATGTCGCATGATGCATGAGAGCT")
```

# Part III: Finding the Replication Origin

## Challenge 1 goals:

1.  Work through an initial window to find frequent k_mers meeting our threshold (t)
2.  Slide the window through the genome, update counts and add frequent k-mers to candidate list

Notes: Windown size L= 5, k=2, t=2

```{r}
initialize_k_mer_dict <- function(k){
  nucleotides <- c("A", "C", "G", "T")

  k_mers_dict <- expand.grid(rep(list(nucleotides), k)) %>%
    unite("k_mers", everything(), remove = TRUE, sep = "") %>%
    unique() %>%
    mutate(count = 0)
  
  return(k_mers_dict)
}

k_mers_dict <- initialize_k_mer_dict(9)
k_mers_dict 
```

```{r}
clump_finding <- function(genomeString , L, t , k ) {
  initial_window <-  str_sub(genome, i, L)
  my_kmer_dict <- initialize_k_mer_dict(k)
  for (i in 1: L-k + 1) {
    curr_kmer <- str_sub(initial_window, i, i + k - 1)
    curr_count <- count_pattern(initial_window, curr_kmer)
    curr_row <- which(my_kmer_dict$kmers == curr_kmer)
    my_kmer_dict$count[curr_row] <- curr_count
  }
  candidates <- my_kmer_dict %>%
    filter(count >= t) %>%
  for (j  in 2:10) {#need to change 10
    
  }
}

```
